<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>5&nbsp; Putting it all together – BIOL90042 R Course</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./stats_primer.html" rel="next">
<link href="./chapter_4.html" rel="prev">
<link href="./favicon.ico" rel="icon">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-a9d6f1d2d70ec6daa267fded9d257be0.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-6402c674ead77b5518d08be0543df564.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/quarto-contrib/alpine-3.12/buttons.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./chapter_5.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Putting it all together</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="./index.html" class="sidebar-logo-link">
      <img src="./WEHI_RGB_logo.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="./">BIOL90042 R Course Book</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setup</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to R</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Working with data - Part 1</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Working with data - Part 2</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_4.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Plotting with ggplot2</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_5.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Putting it all together</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./stats_primer.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Statistics in R Primer</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_6.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">RNA-seq - Part 1</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter_7.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">RNA-seq - Part 2</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./chapter_8.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">8 Machine learning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ch8_section_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Part 1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ch8_section_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Part 2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ch8_section_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Part 3</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./further_reading.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Further reading</span></a>
  </div>
</li>
    </ul>
    </div>
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">In this chapter:</h2>
   
  <ul>
  <li><a href="#introduction-to-the-dataset" id="toc-introduction-to-the-dataset" class="nav-link active" data-scroll-target="#introduction-to-the-dataset"><span class="header-section-number">5.1</span> Introduction to the dataset</a></li>
  <li><a href="#tidy-data" id="toc-tidy-data" class="nav-link" data-scroll-target="#tidy-data"><span class="header-section-number">5.2</span> Tidy Data</a></li>
  <li><a href="#sec-reshaping" id="toc-sec-reshaping" class="nav-link" data-scroll-target="#sec-reshaping"><span class="header-section-number">5.3</span> Reshaping and combining data</a>
  <ul class="collapse">
  <li><a href="#sec-pivot" id="toc-sec-pivot" class="nav-link" data-scroll-target="#sec-pivot"><span class="header-section-number">5.3.1</span> Reshaping data with pivot functions</a></li>
  <li><a href="#separating-data-in-a-column" id="toc-separating-data-in-a-column" class="nav-link" data-scroll-target="#separating-data-in-a-column"><span class="header-section-number">5.3.2</span> Separating data in a column</a></li>
  <li><a href="#summarising-values" id="toc-summarising-values" class="nav-link" data-scroll-target="#summarising-values"><span class="header-section-number">5.3.3</span> Summarising values</a></li>
  <li><a href="#reshaping-for-plotting" id="toc-reshaping-for-plotting" class="nav-link" data-scroll-target="#reshaping-for-plotting"><span class="header-section-number">5.3.4</span> Reshaping for plotting</a></li>
  <li><a href="#sec-join" id="toc-sec-join" class="nav-link" data-scroll-target="#sec-join"><span class="header-section-number">5.3.5</span> Combining data with join functions</a></li>
  </ul></li>
  <li><a href="#visualising-testing-expression-between-strains" id="toc-visualising-testing-expression-between-strains" class="nav-link" data-scroll-target="#visualising-testing-expression-between-strains"><span class="header-section-number">5.4</span> Visualising &amp; testing expression between strains</a>
  <ul class="collapse">
  <li><a href="#simple-linear-model-discrete-x" id="toc-simple-linear-model-discrete-x" class="nav-link" data-scroll-target="#simple-linear-model-discrete-x"><span class="header-section-number">5.4.1</span> Simple linear model (discrete x)</a></li>
  </ul></li>
  <li><a href="#visualising-testing-relationships-between-dosage-and-genes" id="toc-visualising-testing-relationships-between-dosage-and-genes" class="nav-link" data-scroll-target="#visualising-testing-relationships-between-dosage-and-genes"><span class="header-section-number">5.5</span> Visualising &amp; testing relationships between dosage and genes</a>
  <ul class="collapse">
  <li><a href="#simple-linear-model-continuous-x" id="toc-simple-linear-model-continuous-x" class="nav-link" data-scroll-target="#simple-linear-model-continuous-x"><span class="header-section-number">5.5.1</span> Simple linear model (continuous x)</a></li>
  </ul></li>
  <li><a href="#saving-multiple-objects" id="toc-saving-multiple-objects" class="nav-link" data-scroll-target="#saving-multiple-objects"><span class="header-section-number">5.6</span> Saving multiple objects</a></li>
  <li><a href="#another-case-study---who-tuberculosis-data" id="toc-another-case-study---who-tuberculosis-data" class="nav-link" data-scroll-target="#another-case-study---who-tuberculosis-data"><span class="header-section-number">5.7</span> Another case study - WHO tuberculosis data</a>
  <ul class="collapse">
  <li><a href="#string-manupulation" id="toc-string-manupulation" class="nav-link" data-scroll-target="#string-manupulation"><span class="header-section-number">5.7.1</span> String manupulation</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">5.8</span> Summary</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-chapter05" class="quarto-section-identifier"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Putting it all together</span></span></h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i></button></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>In this chapter we will combine all the skills we have learned so far to perform a complete analysis of a small dataset.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Learning Objectives">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Learning Objectives
</div>
</div>
<div class="callout-body-container callout-body">
<p>At the end of this chapter, learners should be able to:</p>
<ol type="1">
<li><p>Describe the key steps in data analysis (exploration, manipulating and plotting)</p></li>
<li><p>Understand how pivot and join functions can be used to reshape and combine entire data frames</p></li>
</ol>
</div>
</div>
<section id="introduction-to-the-dataset" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="introduction-to-the-dataset"><span class="header-section-number">5.1</span> Introduction to the dataset</h2>
<p>In this chapter we will use the mousezempic dosage data <code>m_dose</code> and mouse expression data <code>m_exp</code> data frames, which contain information about the mice including mousezempic dose, and their gene expression levels, respectively.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># read in dosage data</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>m_dose <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/mousezempic_dosage_data.csv"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>m_dose</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in expression data</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>m_exp <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"data/mousezempic_expression_data.tsv"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>m_exp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 453 × 11
   id_num Th_rep1 Th_rep2 Th_rep3 Prlh_rep1 Prlh_rep2 Prlh_rep3 Hprt1_rep1
    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;
 1      1   1431.   2737.   2036.      297.      428.      320.       741.
 2      2   2699.   4180.   3860.      388.      680.      500.       843.
 3      3   3596.   5194.   4128.      201.      346.      266.      1047.
 4      4   2723.   4442.   3728.      317.      576.      478.       541.
 5      5   2197.   3656.   2887.      114.      205.      173.       798.
 6      6   2316.   3949.   3355.      242.      383.      305.       763.
 7      7   1570.   2740.   2129.      294.      489.      427.       796.
 8      8   2683.   4749.   3563.      231.      400.      309.       922.
 9      9   3023.   5170.   3830.      289.      464.      390.       921.
10     10   3254.   5980.   4250.      337.      583.      436.       740.
# ℹ 443 more rows
# ℹ 3 more variables: Hprt1_rep2 &lt;dbl&gt;, Hprt1_rep3 &lt;dbl&gt;, group &lt;chr&gt;</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>We have used <code>read_csv()</code> for the mouse data and <code>read_tsv()</code> for the expression data. These are for reading data separated by commas and tab characters respectively. The <code>readr</code> package also provides <code>read_delim()</code> to let the package guess your delimiter, but if you know the format of your file then it’s good practice to use the appropriate reading function for more predictable behaviour.</p>
</div>
</div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>There’s nothing stopping someone from naming a file <code>file.csv</code> while having tab-separated data inside. This happens quite often in real-world data so it’s a good idea to have a quick look at the data in an text editor before reading it in.</p>
</div>
</div>
</section>
<section id="tidy-data" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="tidy-data"><span class="header-section-number">5.2</span> Tidy Data</h2>
<p>The tidyverse revolves around an important concept called “<a href="https://r4ds.hadley.nz/data-tidy.html">tidy data</a>”. This is a specific representation of tabular data that is considered easy to work with. Tidy data is roughly defined as tabular data that contains:</p>
<ol type="1">
<li>Individual variables in columns</li>
<li>Individual observations in rows</li>
<li>One data value in each cell</li>
</ol>
<p>Having individual variables in columns makes then accessible for performing tidyverse operations like <code>select()</code>, <code>mutate()</code>, <code>filter()</code> and <code>arrange()</code>. If variables were not stored as columns, then these functions would not be able to access them by name.</p>
<p>Having individual observations in rows is important because it associates all variables of each observation with the same row. If the data from one observation is spread across multiple rows then it is easy to incorrect summaries from the <code>summarise()</code> function. When using the <code>filter()</code> function with tidy data, you can expect to keep all the data for an observation or none at all. When the data for observations is split over different rows, it’s possible to unknowingly lose partial data from observations.</p>
<p>Having a single value in each cell makes it possible to perform meaningful computations for the values, for example you cannot take a <code>mean()</code> of a column of values that contain multiple different values.</p>
<p>Although tidy data is the easiest to work with, it’s often necessary to alter the format of your data for plotting or table displays. It’s a good idea to keep your core data in a tidy format and treating plot or table outputs as representations of that tidy data.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Most data you encounter will not be tidy, the first part of data analysis is usually called “data-wranging” and involves tidying up your data so it is easier to use for downstream analysis.</p>
</div>
</div>
</section>
<section id="sec-reshaping" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="sec-reshaping"><span class="header-section-number">5.3</span> Reshaping and combining data</h2>
<p>The <code>filter()</code>, <code>select()</code>, <code>mutate()</code> and <code>summarise()</code> functions we learned last chapter all operate along either the columns or the rows of data. Combining these operations cleverly can answer the majority of questions about your data. However, there are two useful families of functions: <code>pivot</code> for reshaping your data and <code>join</code> for combining your data from shared columns.</p>
<section id="sec-pivot" class="level3" data-number="5.3.1">
<h3 data-number="5.3.1" class="anchored" data-anchor-id="sec-pivot"><span class="header-section-number">5.3.1</span> Reshaping data with pivot functions</h3>
<p>Pivoting is a way to change the shape of your tibble.</p>
<ul>
<li><p>Pivoting longer reshapes the data to transfer data stored in columns into rows, resulting in more rows and fewer columns.</p></li>
<li><p>Pivoting wider is the reverse, moving data from rows into columns.</p></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F10_pivot.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption>Pivot functions allow you to change the structure of your data frame</figcaption>
</figure>
</div>
<p>The <code>pivot_longer()</code> function is used to pivot data from wide to long format, and the <code>pivot_wider()</code> function is used to pivot data from long to wide format.</p>
<section id="pivot-wider" class="level4" data-number="5.3.1.1">
<h4 data-number="5.3.1.1" class="anchored" data-anchor-id="pivot-wider"><span class="header-section-number">5.3.1.1</span> Pivot wider</h4>
<p>A common use case for <code>pivot_wider()</code> is to make a <a href="https://en.wikipedia.org/wiki/Contingency_table">contingency table</a>, which shows the number of observations for each combination of two variables. This is often easier to read than the same information in long format.</p>
<p>For example, let’s say we want to create a table that shows how many mice there are of each strain, in each cage number. We can achieve this in a long format using <code>summarise()</code> as we learned in the previous section:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  cage_number mouse_strain n_mice
  &lt;chr&gt;       &lt;chr&gt;         &lt;int&gt;
1 1A          CD-1             52
2 3E          CD-1             44
3 2B          CD-1             56
4 3E          Black 6         124
5 2B          BALB C           68</code></pre>
</div>
</div>
<p>For the specific task of counting, we can achieve the same effect using the <code>count()</code> tidyverse function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(cage_number, mouse_strain, <span class="at">name =</span> <span class="st">"n_mice"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 3
  cage_number mouse_strain n_mice
  &lt;chr&gt;       &lt;chr&gt;         &lt;int&gt;
1 1A          CD-1             52
2 2B          BALB C           68
3 2B          CD-1             56
4 3E          Black 6         124
5 3E          CD-1             44</code></pre>
</div>
</div>
<p>This summarises by each combination of <code>cage_number</code> and <code>mouse_strain</code>, with the <code>n()</code> function giving the count of data belonging to that combination.</p>
<p>To get a contingency table, we wish to have the information in the <code>mouse_strain</code> column displayed along the column names and the values of <code>n_mice</code> becoming the values of the cells in the new table. Since the goal is to make the table wider, we use the <code>pivot_wider()</code> function. To achieve this, we instruct the <code>pivot_wider()</code> function to take names from the <code>mouse_strain</code> column and the values from the <code>n_mice</code> column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(cage_number, mouse_strain, <span class="at">name =</span> <span class="st">"n_mice"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  cage_number `CD-1` `BALB C` `Black 6`
  &lt;chr&gt;        &lt;int&gt;    &lt;int&gt;     &lt;int&gt;
1 1A              52       NA        NA
2 2B              56       68        NA
3 3E              44       NA       124</code></pre>
</div>
</div>
<p>This has transformed our data into a contingency table, with <code>NA</code> where no data corresponding data exists for the specific <code>cage_number</code> and <code>mouse_strain</code> combination.</p>
<p>We can do the same thing to see how many of each mouse strains is in each of our experiment replicates.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(replicate, mouse_strain, <span class="at">name =</span> <span class="st">"n_mice"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  replicate `BALB C` `Black 6` `CD-1`
  &lt;chr&gt;        &lt;int&gt;     &lt;int&gt;  &lt;int&gt;
1 rep1            26        34     50
2 rep2            18        46     50
3 rep3            24        44     52</code></pre>
</div>
</div>
</section>
<section id="pivot-longer" class="level4" data-number="5.3.1.2">
<h4 data-number="5.3.1.2" class="anchored" data-anchor-id="pivot-longer"><span class="header-section-number">5.3.1.2</span> Pivot longer</h4>
<p>Data can often arrive in the form similar to the contingency table we constructed. Although this data is easy to read, it is difficult to operate on using tidyverse functions because the <code>mouse_strain</code> data is now stored in the column names and not inside a column we can use as a variable. In order to make this kind of data tidy, we use the <code>pivot_longer()</code> function, which will create a pair of columns from the column names and the value of the corresponding cell.</p>
<p>To demonstrate <code>pivot_longer()</code>, we will use the <code>m_exp</code> that we downloaded earlier. This data frame contains the expression levels of two genes (TH and PRLH) suspected to be upregulated in mice taking MouseZempic, as well as one housekeeping gene (HPRT1), all measured in triplicate.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>m_exp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 453 × 11
   id_num Th_rep1 Th_rep2 Th_rep3 Prlh_rep1 Prlh_rep2 Prlh_rep3 Hprt1_rep1
    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;
 1      1   1431.   2737.   2036.      297.      428.      320.       741.
 2      2   2699.   4180.   3860.      388.      680.      500.       843.
 3      3   3596.   5194.   4128.      201.      346.      266.      1047.
 4      4   2723.   4442.   3728.      317.      576.      478.       541.
 5      5   2197.   3656.   2887.      114.      205.      173.       798.
 6      6   2316.   3949.   3355.      242.      383.      305.       763.
 7      7   1570.   2740.   2129.      294.      489.      427.       796.
 8      8   2683.   4749.   3563.      231.      400.      309.       922.
 9      9   3023.   5170.   3830.      289.      464.      390.       921.
10     10   3254.   5980.   4250.      337.      583.      436.       740.
# ℹ 443 more rows
# ℹ 3 more variables: Hprt1_rep2 &lt;dbl&gt;, Hprt1_rep3 &lt;dbl&gt;, group &lt;chr&gt;</code></pre>
</div>
</div>
<p>The data is currently in wide format, with each row representing a different mouse (identified by its <code>id_num</code>) and each column representing a different measurement of a gene. To reshape this data into a long format (where each measurement is contained on a separate row), we can use <code>pivot_longer()</code>, specifying three arguments:</p>
<ol type="1">
<li><code>cols</code>: the columns to pivot from. You can use selection helpers like <code>contains()</code> or <code>starts_with()</code> to easily select multiple columns at once.</li>
<li><code>names_to</code>: the name of a new column that will contain the original column names.</li>
<li><code>values_to</code>: the name of a new column that will contain the values from the original columns.</li>
</ol>
<p>In this particular case here’s what the code would look like:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>m_exp_long <span class="ot">&lt;-</span> m_exp <span class="sc">%&gt;%</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="fu">contains</span>(<span class="st">"_rep"</span>),</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"measurement"</span>,</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">"expression_level"</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>m_exp_long</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4,077 × 4
   id_num group     measurement expression_level
    &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;                  &lt;dbl&gt;
 1      1 treatment Th_rep1                1431.
 2      1 treatment Th_rep2                2737.
 3      1 treatment Th_rep3                2036.
 4      1 treatment Prlh_rep1               297.
 5      1 treatment Prlh_rep2               428.
 6      1 treatment Prlh_rep3               320.
 7      1 treatment Hprt1_rep1              741.
 8      1 treatment Hprt1_rep2             1300.
 9      1 treatment Hprt1_rep3              988.
10      2 treatment Th_rep1                2699.
# ℹ 4,067 more rows</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Perplexed by pivoting?">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Perplexed by pivoting?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Pivoting can be a bit tricky to get your head around! Often when you’re doing analysis, you’ll run into the problem of knowing that you need to pivot, but not knowing exactly what arguments to use. In these cases, it can be helpful to look at examples online, <a href="https://r4ds.hadley.nz/data-tidy.html#sec-pivoting">like those in the R for Data Science book</a>.</p>
</div>
</div>
</section>
</section>
<section id="separating-data-in-a-column" class="level3" data-number="5.3.2">
<h3 data-number="5.3.2" class="anchored" data-anchor-id="separating-data-in-a-column"><span class="header-section-number">5.3.2</span> Separating data in a column</h3>
<p>When we look at the <code>measurement</code> column we see that it contains two pieces of information. The gene being measured and the replicate number separated by “_”. We can use the <code>separate()</code> function to split this data into individual columns so that one column does not contain multiple variables.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>m_exp_separate <span class="ot">&lt;-</span> m_exp_long <span class="sc">%&gt;%</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(measurement, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"gene"</span>, <span class="st">"replicate"</span>), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>m_exp_separate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4,077 × 5
   id_num group     gene  replicate expression_level
    &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt;                &lt;dbl&gt;
 1      1 treatment Th    rep1                 1431.
 2      1 treatment Th    rep2                 2737.
 3      1 treatment Th    rep3                 2036.
 4      1 treatment Prlh  rep1                  297.
 5      1 treatment Prlh  rep2                  428.
 6      1 treatment Prlh  rep3                  320.
 7      1 treatment Hprt1 rep1                  741.
 8      1 treatment Hprt1 rep2                 1300.
 9      1 treatment Hprt1 rep3                  988.
10      2 treatment Th    rep1                 2699.
# ℹ 4,067 more rows</code></pre>
</div>
</div>
</section>
<section id="summarising-values" class="level3" data-number="5.3.3">
<h3 data-number="5.3.3" class="anchored" data-anchor-id="summarising-values"><span class="header-section-number">5.3.3</span> Summarising values</h3>
<p>Now suppose we wanted to combine the gene expression data across replicates by adding them up, we can use the <code>summarise()</code> function to do so.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>m_exp_rep_summed <span class="ot">&lt;-</span> m_exp_separate <span class="sc">%&gt;%</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">expression_level =</span> <span class="fu">sum</span>(expression_level),</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(id_num, group, gene)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>m_exp_rep_summed</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,359 × 4
   id_num group     gene  expression_level
    &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;            &lt;dbl&gt;
 1      1 treatment Th               6204.
 2      1 treatment Prlh             1045.
 3      1 treatment Hprt1            3029.
 4      2 treatment Th              10738.
 5      2 treatment Prlh             1568.
 6      2 treatment Hprt1            3669.
 7      3 treatment Th              12918.
 8      3 treatment Prlh              813.
 9      3 treatment Hprt1            4297.
10      4 treatment Th              10892.
# ℹ 1,349 more rows</code></pre>
</div>
</div>
<p>We may also want to get the average expression level of each gene within each experimental group and plot it as a bar chart.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>m_exp_rep_summed <span class="sc">%&gt;%</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">expression_level =</span> <span class="fu">mean</span>(expression_level),</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(group, gene)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene, <span class="at">y =</span> expression_level, <span class="at">fill =</span> group)) <span class="sc">+</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="at">position =</span> <span class="st">"dodge"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="chapter_5_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>We see that all 3 genes are slightly higher in the treatment group.</p>
</section>
<section id="reshaping-for-plotting" class="level3" data-number="5.3.4">
<h3 data-number="5.3.4" class="anchored" data-anchor-id="reshaping-for-plotting"><span class="header-section-number">5.3.4</span> Reshaping for plotting</h3>
<p>Suppose we wanted to plot the value of two genes against each other. We would need the expression values of genes to be in individual columns. However we have lost this structure after using <code>pivot_longer()</code>. We can restore it using <code>pivot_wider()</code> after we have summarised the replicates.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>m_exp_rep_summed_wide <span class="ot">&lt;-</span> m_exp_rep_summed <span class="sc">%&gt;%</span> </span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>( <span class="at">names_from =</span> <span class="st">"gene"</span>, <span class="at">values_from =</span> <span class="st">"expression_level"</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>m_exp_rep_summed_wide</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 453 × 5
   id_num group         Th  Prlh Hprt1
    &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1      1 treatment  6204. 1045. 3029.
 2      2 treatment 10738. 1568. 3669.
 3      3 treatment 12918.  813. 4297.
 4      4 treatment 10892. 1371. 2213.
 5      5 treatment  8740.  491. 3159.
 6      6 treatment  9620.  930. 2875.
 7      7 treatment  6440. 1211. 3339.
 8      8 treatment 10996.  939. 3854.
 9      9 treatment 12022. 1144. 3781.
10     10 treatment 13485. 1356. 2848.
# ℹ 443 more rows</code></pre>
</div>
</div>
<p>With the replicate aggregated expression levels in individual columns, we can now plot the gene expression values against each other.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>m_exp_rep_summed_wide <span class="sc">%&gt;%</span> <span class="fu">ggplot</span>( <span class="fu">aes</span>(<span class="at">x =</span> Th, <span class="at">y =</span> Prlh)) <span class="sc">+</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="chapter_5_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Practice exercise">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Practice exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>Make all the pair-wise scatter plots and assemble them together using <code>patchwork</code>. There should be plots of:</p>
<p>Th vs Prlh</p>
<p>Th vs Hprt1</p>
<p>Prlh vs Hprt1</p>
</div>
</div>
</section>
<section id="sec-join" class="level3" data-number="5.3.5">
<h3 data-number="5.3.5" class="anchored" data-anchor-id="sec-join"><span class="header-section-number">5.3.5</span> Combining data with join functions</h3>
<p>Data analysis typically involves multiple tables of data. Often there will be tables that are contain related information that must be combined to answer the questions you’re interested in. Tables that are related to each other tend to have one or more columns in common, and are referred to as “relational data”. Combining relational data in useful ways requires the <code>join</code> family of functions. In general joins can accomplish two tasks</p>
<ul>
<li>Add new variables to an existing table containing additional information.</li>
<li>Filter observations in one table based on whether or not they match observations in another table.</li>
<li>Both of the above at the same time.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/S02F11_left_join.png" class="img-fluid figure-img"></p>
<figcaption>The left_join() function allows you to combine two data frames</figcaption>
</figure>
</div>
<p>A common and basic join is the <code>left_join()</code>. It takes two data frame as arguments and optionally a vector of common columns to perform the join on. The reason it’s called a left-join because it retains all rows from the left data frame while adding on columns from the right data frame only when the data in the designed joining column(s) match.</p>
<p>For example, we can join the <code>m_dose</code> data frame with the <code>m_exp_rep_summed_wide</code> data frame based on the <code>id_num</code> column, which is common to both data frames:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>(m_dose, m_exp_rep_summed_wide, <span class="at">by =</span> <span class="st">"id_num"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 13
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 7 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   id_num &lt;dbl&gt;, group &lt;chr&gt;, Th &lt;dbl&gt;, Prlh &lt;dbl&gt;, Hprt1 &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The <code>m_dose</code> contains information about characteristics of each mouse, while the <code>m_exp_rep_summed_wide</code> contains the replicate-summed gene expressions of the mice. Each data frame has an <code>id_num</code> column that relates the data from the two data frames together, and we have joined them into one table that contains both the data about the mice as well as the expression of their genes.</p>
<p>With the <code>left_join()</code>, if there is a <code>id_num</code> value that exists in the right data frame but not found in the left, then that data will not be present in the joined table. For any <code>id_num</code> that appears only in the left but not the right data frame, the values in the newly created columns will be <code>NA</code>.</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>If the values from the left data frame matches to multiple rows of the column in the right data frame, the <code>left_join()</code> will duplicate the data from the left data frame for each match to the right. This can cause issues with downstream summarisation if not carefully considered.</p>
</div>
</div>
<section id="joining-with-mismatched-column-names" class="level4" data-number="5.3.5.1">
<h4 data-number="5.3.5.1" class="anchored" data-anchor-id="joining-with-mismatched-column-names"><span class="header-section-number">5.3.5.1</span> Joining with mismatched column names</h4>
<p>Often the column containing the matching information has different names in different data frames. For example what might be called “id_num” in one data frame could also be called “mouse_id” in another data frame. In those cases the <code>by</code> argument of <code>left_join()</code> can be formatted to let the function know which column in the left data frame matches to which column on the right.</p>
<p>We will demonstrate this by renaming the <code>id_num</code> column in the <code>m_dose</code> to <code>mouse_id</code> and using it to perform the join instead. We will use the <code>join_by(mouse_id == id_num)</code> helper function for the <code>by</code> argument to specify the different columns we wish to join by.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rename id_num to mouse_id</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>m_dose_data_new <span class="ot">&lt;-</span> m_dose <span class="sc">%&gt;%</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">mouse_id =</span> id_num)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>m_dose_data_new</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 344 × 9
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                  NA    rep1      &lt;NA&gt;      NA      
 5 CD-1         1A                   3.45 rep1      female     0.00193
 6 CD-1         1A                   3.65 rep1      male       0.0019 
 7 CD-1         1A                   3.62 rep1      female     0.00181
 8 CD-1         1A                   4.68 rep1      male       0.00195
 9 CD-1         1A                   3.48 rep1      &lt;NA&gt;       0.00193
10 CD-1         1A                   4.25 rep1      &lt;NA&gt;       0.0019 
# ℹ 334 more rows
# ℹ 3 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   mouse_id &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># perform left join by matching mouse_id of the left data frame to id_num of the right data frame</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>m_joined_data <span class="ot">&lt;-</span> <span class="fu">left_join</span>(</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  m_dose_data_new, m_exp_rep_summed_wide,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">by =</span> <span class="fu">join_by</span>(mouse_id <span class="sc">==</span> id_num)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>() <span class="co"># keep only rows without NA</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>m_joined_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 333 × 13
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.8  rep1      female     0.00186
 3 CD-1         1A                   3.25 rep1      female     0.00195
 4 CD-1         1A                   3.45 rep1      female     0.00193
 5 CD-1         1A                   3.65 rep1      male       0.0019 
 6 CD-1         1A                   3.62 rep1      female     0.00181
 7 CD-1         1A                   4.68 rep1      male       0.00195
 8 CD-1         1A                   3.2  rep1      female     0.00182
 9 CD-1         1A                   3.8  rep1      male       0.00191
10 CD-1         1A                   4.4  rep1      male       0.00198
# ℹ 323 more rows
# ℹ 7 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   mouse_id &lt;dbl&gt;, group &lt;chr&gt;, Th &lt;dbl&gt;, Prlh &lt;dbl&gt;, Hprt1 &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>When joining data frames with mismatched column names, the column name from the left data frame argument will be retained for the joined result.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>An older syntax exists for the <code>by</code> argument that would look like <code>by = c("mouse_id" = "id_num")</code>. This does the same thing as what is shown above but the documentation now recommends the new syntax which allows for flexible options like joining to the closest.</p>
</div>
</div>
</section>
</section>
</section>
<section id="visualising-testing-expression-between-strains" class="level2" data-number="5.4">
<h2 data-number="5.4" class="anchored" data-anchor-id="visualising-testing-expression-between-strains"><span class="header-section-number">5.4</span> Visualising &amp; testing expression between strains</h2>
<p>With the joined data we now have more information about the mice in addition to the expression data. So we can check if there is any significant difference in gene expressions between mouse strains.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>m_joined_data <span class="sc">%&gt;%</span> </span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>( <span class="fu">aes</span>(<span class="at">x =</span> mouse_strain, <span class="at">y =</span> Th, <span class="at">fill =</span> mouse_strain)) <span class="sc">+</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="chapter_5_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>If we wanted to plot all 3 genes, one way is to make 3 different plots, but we can also use facets. However to use facets we would need to have the faceting variable in a column, and currently our gene names are in the column names. So we must use <code>pivot_longer()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>m_joined_data_long <span class="ot">&lt;-</span> m_joined_data <span class="sc">%&gt;%</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(Th, Prlh, Hprt1), <span class="at">names_to =</span> <span class="st">"gene"</span>, <span class="at">values_to =</span> <span class="st">"expression"</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>m_joined_data_long</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 999 × 12
   mouse_strain cage_number weight_lost_g replicate sex    drug_dose_g
   &lt;chr&gt;        &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;
 1 CD-1         1A                   3.75 rep1      male       0.00181
 2 CD-1         1A                   3.75 rep1      male       0.00181
 3 CD-1         1A                   3.75 rep1      male       0.00181
 4 CD-1         1A                   3.8  rep1      female     0.00186
 5 CD-1         1A                   3.8  rep1      female     0.00186
 6 CD-1         1A                   3.8  rep1      female     0.00186
 7 CD-1         1A                   3.25 rep1      female     0.00195
 8 CD-1         1A                   3.25 rep1      female     0.00195
 9 CD-1         1A                   3.25 rep1      female     0.00195
10 CD-1         1A                   3.45 rep1      female     0.00193
# ℹ 989 more rows
# ℹ 6 more variables: tail_length_mm &lt;dbl&gt;, initial_weight_g &lt;dbl&gt;,
#   mouse_id &lt;dbl&gt;, group &lt;chr&gt;, gene &lt;chr&gt;, expression &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Now that the gene variable is in a column, we can use this information to create our faceted plot. We use <code>facet_wrap()</code> with the <code>scales = "free_y"</code> argument so each gene can have its own y-axis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>m_joined_data_long <span class="sc">%&gt;%</span> </span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>( <span class="fu">aes</span>(<span class="at">x =</span> mouse_strain, <span class="at">y =</span> expression, <span class="at">fill =</span> mouse_strain)) <span class="sc">+</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>() <span class="sc">+</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>gene, <span class="at">scales =</span> <span class="st">"free_y"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="chapter_5_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>There doesn’t appear to be clear differences in the expression of these three genes between the haplotypes.</p>
<section id="simple-linear-model-discrete-x" class="level3" data-number="5.4.1">
<h3 data-number="5.4.1" class="anchored" data-anchor-id="simple-linear-model-discrete-x"><span class="header-section-number">5.4.1</span> Simple linear model (discrete x)</h3>
<p>To fit a line through the gene expression values of each mouse genotype, and test for significance of the x~y relationship, we can use a handy function <a href="https://github.com/bansell/tidyrstats?tab=readme-ov-file#lm_test"><code>lm_test()</code></a> from the <code>tidyrstats</code> library.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyrstats)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This function takes a long-format data frame as input, and the formula for the linear model we want to fit. (We will read more about linear models in later chapters). For this test, for each gene, we want to know whether the response (gene expression ; aka ‘y’), changes significantly with the predictor (mouse strain ; aka ‘x’). The model formula will be <code>y ~ x1</code>, that is <code>expression ~ mouse_strain</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>m_joined_data_long <span class="sc">%&gt;%</span> </span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gene) <span class="sc">%&gt;%</span> </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lm_test</span>(expression <span class="sc">~</span> mouse_strain) <span class="sc">%&gt;%</span> </span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(term<span class="sc">!=</span><span class="st">'intercept'</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(gene,term)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Results for linear model: expression ~ mouse_strain</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 6
  gene  term                estimate std.error statistic p.value
  &lt;chr&gt; &lt;chr&gt;                  &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;
1 Hprt1 mouse_strainBlack 6     3.31      68.4    0.0485   0.961
2 Hprt1 mouse_strainCD-1       48.2       66.0    0.730    0.466
3 Prlh  mouse_strainBlack 6    10.2       42.3    0.240    0.810
4 Prlh  mouse_strainCD-1      -19.8       40.9   -0.485    0.628
5 Th    mouse_strainBlack 6   666.       557.     1.20     0.233
6 Th    mouse_strainCD-1      697.       538.     1.29     0.196</code></pre>
</div>
</div>
<p>Our intuition from the violin plots is correct. None of the genes show significant differences in expression between the Black6 vs the reference (BALBC), nor the CD-1 vs the reference. Note that all of the P values are &gt; 0.05, meaning we dont have enough evidence to reject the null hypothesis.</p>
</section>
</section>
<section id="visualising-testing-relationships-between-dosage-and-genes" class="level2" data-number="5.5">
<h2 data-number="5.5" class="anchored" data-anchor-id="visualising-testing-relationships-between-dosage-and-genes"><span class="header-section-number">5.5</span> Visualising &amp; testing relationships between dosage and genes</h2>
<p>With the joined data frame we can also visualise the relationship between the drug dosage and the genes, separated by mouse strain. We do this by using our <code>m_joined_data_long</code> to plot with <code>facet_grid()</code> specifying the layout of our grid.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>m_joined_data_long <span class="sc">%&gt;%</span> </span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> drug_dose_g, <span class="at">y =</span> expression, <span class="at">color =</span> mouse_strain)) <span class="sc">+</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>) <span class="sc">+</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="at">cols =</span> <span class="fu">vars</span>(mouse_strain), <span class="at">rows =</span> <span class="fu">vars</span>(gene), <span class="at">scales =</span> <span class="st">"free"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="chapter_5_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
<p>Although very minor, there appears to be some effect of dosage on gene expression that is strain specific.</p>
<section id="simple-linear-model-continuous-x" class="level3" data-number="5.5.1">
<h3 data-number="5.5.1" class="anchored" data-anchor-id="simple-linear-model-continuous-x"><span class="header-section-number">5.5.1</span> Simple linear model (continuous x)</h3>
<p>Again, we can test this using <code>lm_test()</code>, where y is <code>expression</code>, and x is now the continuous variable <code>drug_dose_g</code>. We are testing the association for each gene in each mouse strain, by grouping the input data before <code>lm_test()</code>. Note this is different from correcting for gene and mouse strain as covariates, because the model only tests one gene-strain combination at a time.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>m_joined_data_long <span class="sc">%&gt;%</span> </span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(mouse_strain, gene) <span class="sc">%&gt;%</span> </span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lm_test</span>(expression <span class="sc">~</span> drug_dose_g) <span class="sc">%&gt;%</span> </span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(term<span class="sc">!=</span><span class="st">'intercept'</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, gene, p.value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Results for linear model: expression ~ drug_dose_g</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 9 × 7
  mouse_strain gene  term         estimate std.error statistic p.value
  &lt;chr&gt;        &lt;chr&gt; &lt;chr&gt;           &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;
1 BALB C       Hprt1 drug_dose_g -1531022.   862835.    -1.77  0.0806 
2 BALB C       Prlh  drug_dose_g  -319172.   405419.    -0.787 0.434  
3 BALB C       Th    drug_dose_g  4018858.  6176909.     0.651 0.518  
4 Black 6      Hprt1 drug_dose_g -1561066.   582482.    -2.68  0.00842
5 Black 6      Prlh  drug_dose_g  -391751.   383201.    -1.02  0.309  
6 Black 6      Th    drug_dose_g -4385855.  5391197.    -0.814 0.418  
7 CD-1         Hprt1 drug_dose_g   377655.   556772.     0.678 0.499  
8 CD-1         Prlh  drug_dose_g   478143.   379983.     1.26  0.210  
9 CD-1         Th    drug_dose_g  5299211.  4503093.     1.18  0.241  </code></pre>
</div>
</div>
<p>Here we can see, in fact there is a significant association between Mousezempic drug dose and gene expression: within the Black 6 strain, for the Hprt1 gene.</p>
<div class="callout callout-style-default callout-note callout-titled" title="Stating statistical findings">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Stating statistical findings
</div>
</div>
<div class="callout-body-container callout-body">
<p>From our summary table above, can you finish this sentence which is the formal statement of results from a linear model?</p>
<p><em>“In the Black 6 mouse strain, for every 1 gram increase in drug dose, on average, Hprt1 gene expression …”</em></p>
</div>
</div>
<p>You can read and work through more <code>lm_test()</code> examples in the <a href="./stats_primer.html">Statistics in R Primer</a>.</p>
</section>
</section>
<section id="saving-multiple-objects" class="level2" data-number="5.6">
<h2 data-number="5.6" class="anchored" data-anchor-id="saving-multiple-objects"><span class="header-section-number">5.6</span> Saving multiple objects</h2>
<p>We can save our newly transformed and joined data frames for easy access using the <code>save()</code> command, which takes any number of environment objects, and a new <code>Rdata</code> format file name as arguments:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  m_exp_rep_summed,</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  m_exp_rep_summed_wide,</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>  m_joined_data_long,</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">'data_processed/mousezempic_cleaned_joined.Rdata'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="another-case-study---who-tuberculosis-data" class="level2" data-number="5.7">
<h2 data-number="5.7" class="anchored" data-anchor-id="another-case-study---who-tuberculosis-data"><span class="header-section-number">5.7</span> Another case study - WHO tuberculosis data</h2>
<p>We can look at another dataset containing tuberculosis (TB) cases broken down by year, country, age, gender, and diagnosis method. The data comes from the <a href="https://www.who.int/teams/global-tuberculosis-programme/data">2014 World Health Organization Global Tuberculosis Report</a> and is typical of many real-life datasets in terms of structure. There are redundant columns, odd variable codes and missing values.</p>
<p>We want to tidy it up using the tools we have learned in order to try to extract some information from this data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>who</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 7,240 × 60
   country  iso2  iso3   year new_sp_m014 new_sp_m1524 new_sp_m2534 new_sp_m3544
   &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;
 1 Afghani… AF    AFG    1980          NA           NA           NA           NA
 2 Afghani… AF    AFG    1981          NA           NA           NA           NA
 3 Afghani… AF    AFG    1982          NA           NA           NA           NA
 4 Afghani… AF    AFG    1983          NA           NA           NA           NA
 5 Afghani… AF    AFG    1984          NA           NA           NA           NA
 6 Afghani… AF    AFG    1985          NA           NA           NA           NA
 7 Afghani… AF    AFG    1986          NA           NA           NA           NA
 8 Afghani… AF    AFG    1987          NA           NA           NA           NA
 9 Afghani… AF    AFG    1988          NA           NA           NA           NA
10 Afghani… AF    AFG    1989          NA           NA           NA           NA
# ℹ 7,230 more rows
# ℹ 52 more variables: new_sp_m4554 &lt;dbl&gt;, new_sp_m5564 &lt;dbl&gt;,
#   new_sp_m65 &lt;dbl&gt;, new_sp_f014 &lt;dbl&gt;, new_sp_f1524 &lt;dbl&gt;,
#   new_sp_f2534 &lt;dbl&gt;, new_sp_f3544 &lt;dbl&gt;, new_sp_f4554 &lt;dbl&gt;,
#   new_sp_f5564 &lt;dbl&gt;, new_sp_f65 &lt;dbl&gt;, new_sn_m014 &lt;dbl&gt;,
#   new_sn_m1524 &lt;dbl&gt;, new_sn_m2534 &lt;dbl&gt;, new_sn_m3544 &lt;dbl&gt;,
#   new_sn_m4554 &lt;dbl&gt;, new_sn_m5564 &lt;dbl&gt;, new_sn_m65 &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>The first thing we notice is that <code>country</code>, <code>iso2</code> and <code>iso3</code> all seem to be different encodings for the country. The <code>year</code> column is fine as a variable but the rest of the columns seem appear to be coding values in the column names. There are numeric values in the cells, and since we know this dataset is supposed to contain case numbers, and no column clearly denotes this, we suspect that the values in the columns are case numbers with the column names being descriptive of the cases.</p>
<p>Since we believe that many of these column names are in fact encoding data, we want to <code>pivot_longer()</code> to put that information in a column for further processing.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pivot longer with columns 5 to 60, drop NA values</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>who_long <span class="ot">&lt;-</span> who <span class="sc">%&gt;%</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="dv">5</span><span class="sc">:</span><span class="dv">60</span>, <span class="co"># alternatively contains("_")</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"description"</span>,</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">"cases"</span>,</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_drop_na =</span> <span class="cn">TRUE</span></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>who_long</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 76,046 × 6
   country     iso2  iso3   year description  cases
   &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt;
 1 Afghanistan AF    AFG    1997 new_sp_m014      0
 2 Afghanistan AF    AFG    1997 new_sp_m1524    10
 3 Afghanistan AF    AFG    1997 new_sp_m2534     6
 4 Afghanistan AF    AFG    1997 new_sp_m3544     3
 5 Afghanistan AF    AFG    1997 new_sp_m4554     5
 6 Afghanistan AF    AFG    1997 new_sp_m5564     2
 7 Afghanistan AF    AFG    1997 new_sp_m65       0
 8 Afghanistan AF    AFG    1997 new_sp_f014      5
 9 Afghanistan AF    AFG    1997 new_sp_f1524    38
10 Afghanistan AF    AFG    1997 new_sp_f2534    36
# ℹ 76,036 more rows</code></pre>
</div>
</div>
<p>It doesn’t look like the values in <code>description</code> are unique, so we can try to count them to get a sense of what we’re dealing with. It will show us the unique values as well as the counts for each value.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>who_long <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(description)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 56 × 2
   description      n
   &lt;chr&gt;        &lt;int&gt;
 1 new_ep_f014   1032
 2 new_ep_f1524  1021
 3 new_ep_f2534  1021
 4 new_ep_f3544  1021
 5 new_ep_f4554  1017
 6 new_ep_f5564  1017
 7 new_ep_f65    1014
 8 new_ep_m014   1038
 9 new_ep_m1524  1026
10 new_ep_m2534  1020
# ℹ 46 more rows</code></pre>
</div>
</div>
<p>Some patterns emerge, and it maybe be possible to guess at exactly what these values mean. For example we see the presence of “m” and “f” in the encoding which probably encodes sex. Luckily we have a data dictionary ready from the source that tells us what this means.</p>
<ol type="1">
<li>The first part tells us if the data is new or old, in this case all data is new.</li>
<li>The next part describe the type of TB:</li>
</ol>
<ul>
<li><code>rel</code> for relapse</li>
<li><code>ep</code> for extrapulmonary</li>
<li><code>sn</code> for pulmonary TB that cannot be diagnosed with a pulmonary smear (smear negative)</li>
<li><code>sp</code> for pulmonary TB that can be diagnosed with a smear (smear positive)</li>
</ul>
<ol start="3" type="1">
<li>The sex of the patient. Males and females denoted by <code>m</code> and <code>f</code> respectively.</li>
<li>The 7 age groups the data is divided into:</li>
</ol>
<ul>
<li><code>014</code>: 0 - 14 years old</li>
<li><code>1524</code>: 15 - 24 years old</li>
<li><code>2534</code>: 25 - 34 years old</li>
<li><code>3544</code>: 35 - 44 years old</li>
<li><code>4554</code>: 45 - 54 years old</li>
<li><code>5564</code>: 55 - 64 years old</li>
<li><code>65</code>: 65 or older</li>
</ul>
<section id="string-manupulation" class="level3" data-number="5.7.1">
<h3 data-number="5.7.1" class="anchored" data-anchor-id="string-manupulation"><span class="header-section-number">5.7.1</span> String manupulation</h3>
<p>Armed with that knowledge we can begin to tidy up our dataset! The first thing we will do is tidy up the <code>description</code> column, we will do this with some string manipulation functions from the tidyverse. These functions help us manipulate strings programatically and are essential in dealing with real-world data.</p>
<p>Some useful functions for string manipulation include:</p>
<ul>
<li><code>str_replace(string, pattern, replacement)</code>: replaces the pattern in the string with the replacement.</li>
<li><code>str_extract(string, pattern)</code>: extracts the part of the string matching the pattern.</li>
<li><code>str_remove(string, pattern)</code>: removes the part of the string matching the pattern.</li>
</ul>
<p>The full list of functions can be found in the <a href="https://stringr.tidyverse.org/reference/index.html">stringr package documentation</a> and patterns can be written in very powerful ways with <a href="https://r4ds.had.co.nz/strings.html#matching-patterns-with-regular-expressions">regular expressions</a>. For our purposes we will just use simply fixed patterns.</p>
<p>We saw previously that we can use <code>separate()</code> to break apart on column into multiple columns. However, we see there is a problem in this dataset where the cases that are marked as relapse are encoded <code>newrel</code> which doesn’t leave us with a separator to use. We can remedy this by using a combination of <code>mutate()</code> and <code>str_replace()</code> to replace all instances of <code>newrel</code> with <code>new_rel</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>who_long <span class="ot">&lt;-</span> who_long <span class="sc">%&gt;%</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">description =</span> <span class="fu">str_replace</span>(description, <span class="st">"newrel"</span>, <span class="st">"new_rel"</span>))</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>who_long</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 76,046 × 6
   country     iso2  iso3   year description  cases
   &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt;
 1 Afghanistan AF    AFG    1997 new_sp_m014      0
 2 Afghanistan AF    AFG    1997 new_sp_m1524    10
 3 Afghanistan AF    AFG    1997 new_sp_m2534     6
 4 Afghanistan AF    AFG    1997 new_sp_m3544     3
 5 Afghanistan AF    AFG    1997 new_sp_m4554     5
 6 Afghanistan AF    AFG    1997 new_sp_m5564     2
 7 Afghanistan AF    AFG    1997 new_sp_m65       0
 8 Afghanistan AF    AFG    1997 new_sp_f014      5
 9 Afghanistan AF    AFG    1997 new_sp_f1524    38
10 Afghanistan AF    AFG    1997 new_sp_f2534    36
# ℹ 76,036 more rows</code></pre>
</div>
</div>
<p>Since every single observation contains only new cases and no old cases, the <code>new_</code> portion of the encoding is redundant. So we can remove it using <code>str_remove()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>who_long <span class="ot">&lt;-</span> who_long <span class="sc">%&gt;%</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">description =</span> <span class="fu">str_remove</span>(description, <span class="st">"new_"</span>))</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>who_long</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 76,046 × 6
   country     iso2  iso3   year description cases
   &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;       &lt;dbl&gt;
 1 Afghanistan AF    AFG    1997 sp_m014         0
 2 Afghanistan AF    AFG    1997 sp_m1524       10
 3 Afghanistan AF    AFG    1997 sp_m2534        6
 4 Afghanistan AF    AFG    1997 sp_m3544        3
 5 Afghanistan AF    AFG    1997 sp_m4554        5
 6 Afghanistan AF    AFG    1997 sp_m5564        2
 7 Afghanistan AF    AFG    1997 sp_m65          0
 8 Afghanistan AF    AFG    1997 sp_f014         5
 9 Afghanistan AF    AFG    1997 sp_f1524       38
10 Afghanistan AF    AFG    1997 sp_f2534       36
# ℹ 76,036 more rows</code></pre>
</div>
</div>
<p>Now one last problem remains, there is no separator between the sex and age group. So we can once again use <code>str_replace()</code> to edit in an underscore. Note that we use <code>"_f"</code> and <code>"_m"</code> as our patterns as a precaution in case the letters naturally occur elsewhere in the encoding.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>who_long <span class="ot">&lt;-</span> who_long <span class="sc">%&gt;%</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">description =</span> <span class="fu">str_replace</span>(description, <span class="st">"_f"</span>, <span class="st">"_f_"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">description =</span> <span class="fu">str_replace</span>(description, <span class="st">"_m"</span>, <span class="st">"_m_"</span>))</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>who_long</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 76,046 × 6
   country     iso2  iso3   year description cases
   &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;       &lt;dbl&gt;
 1 Afghanistan AF    AFG    1997 sp_m_014        0
 2 Afghanistan AF    AFG    1997 sp_m_1524      10
 3 Afghanistan AF    AFG    1997 sp_m_2534       6
 4 Afghanistan AF    AFG    1997 sp_m_3544       3
 5 Afghanistan AF    AFG    1997 sp_m_4554       5
 6 Afghanistan AF    AFG    1997 sp_m_5564       2
 7 Afghanistan AF    AFG    1997 sp_m_65         0
 8 Afghanistan AF    AFG    1997 sp_f_014        5
 9 Afghanistan AF    AFG    1997 sp_f_1524      38
10 Afghanistan AF    AFG    1997 sp_f_2534      36
# ℹ 76,036 more rows</code></pre>
</div>
</div>
<p>With that we are finally able to separate the data into individual columns with <code>separate()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>who_separated <span class="ot">&lt;-</span> who_long <span class="sc">%&gt;%</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(description, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"type"</span>, <span class="st">"sex"</span>, <span class="st">"age_group"</span>), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>who_separated</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 76,046 × 8
   country     iso2  iso3   year type  sex   age_group cases
   &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;
 1 Afghanistan AF    AFG    1997 sp    m     014           0
 2 Afghanistan AF    AFG    1997 sp    m     1524         10
 3 Afghanistan AF    AFG    1997 sp    m     2534          6
 4 Afghanistan AF    AFG    1997 sp    m     3544          3
 5 Afghanistan AF    AFG    1997 sp    m     4554          5
 6 Afghanistan AF    AFG    1997 sp    m     5564          2
 7 Afghanistan AF    AFG    1997 sp    m     65            0
 8 Afghanistan AF    AFG    1997 sp    f     014           5
 9 Afghanistan AF    AFG    1997 sp    f     1524         38
10 Afghanistan AF    AFG    1997 sp    f     2534         36
# ℹ 76,036 more rows</code></pre>
</div>
</div>
<p>Because the columns <code>iso2</code> and <code>iso3</code> contain redundant information about the country, we can remove them, but also retain them for reference as we may want to know what the short forms of the countries are when relating it other tables. To do so we can <code>select()</code> the columns containing this information, then use <code>distinct()</code> to keep only unique rows.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>country_codes <span class="ot">&lt;-</span> who_separated <span class="sc">%&gt;%</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(country, iso2, iso3) <span class="sc">%&gt;%</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>()</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>country_codes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 219 × 3
   country             iso2  iso3 
   &lt;chr&gt;               &lt;chr&gt; &lt;chr&gt;
 1 Afghanistan         AF    AFG  
 2 Albania             AL    ALB  
 3 Algeria             DZ    DZA  
 4 American Samoa      AS    ASM  
 5 Andorra             AD    AND  
 6 Angola              AO    AGO  
 7 Anguilla            AI    AIA  
 8 Antigua and Barbuda AG    ATG  
 9 Argentina           AR    ARG  
10 Armenia             AM    ARM  
# ℹ 209 more rows</code></pre>
</div>
</div>
<p>With that we can safely remove the redundant information from our data and end up with a tidy data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>who_tidy <span class="ot">&lt;-</span> who_separated <span class="sc">%&gt;%</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>iso2, <span class="sc">-</span>iso3)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>who_tidy</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 76,046 × 6
   country      year type  sex   age_group cases
   &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;
 1 Afghanistan  1997 sp    m     014           0
 2 Afghanistan  1997 sp    m     1524         10
 3 Afghanistan  1997 sp    m     2534          6
 4 Afghanistan  1997 sp    m     3544          3
 5 Afghanistan  1997 sp    m     4554          5
 6 Afghanistan  1997 sp    m     5564          2
 7 Afghanistan  1997 sp    m     65            0
 8 Afghanistan  1997 sp    f     014           5
 9 Afghanistan  1997 sp    f     1524         38
10 Afghanistan  1997 sp    f     2534         36
# ℹ 76,036 more rows</code></pre>
</div>
</div>
<p>From there we can easily visualise various aspects of the data for analysis. For example we can see what country had the highest number of total cases in the year 2010.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>who_tidy <span class="sc">%&gt;%</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(year <span class="sc">==</span> <span class="dv">2010</span>) <span class="sc">%&gt;%</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">total_cases =</span> <span class="fu">sum</span>(cases), <span class="at">.by =</span> country) <span class="sc">%&gt;%</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(total_cases))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 204 × 2
   country                               total_cases
   &lt;chr&gt;                                       &lt;dbl&gt;
 1 China                                      869092
 2 India                                      630164
 3 South Africa                               364992
 4 Indonesia                                  296272
 5 Bangladesh                                 150856
 6 Pakistan                                   124455
 7 Russian Federation                         102823
 8 Philippines                                 89198
 9 Kenya                                       85484
10 Democratic People's Republic of Korea       81240
# ℹ 194 more rows</code></pre>
</div>
</div>
<p>We can extract only the data from the top 10 countries based in 2010 cases and plot how their total case numbers between the years 2000 and 2010.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>top_10_countries <span class="ot">&lt;-</span> who_tidy <span class="sc">%&gt;%</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(year <span class="sc">==</span> <span class="dv">2010</span>) <span class="sc">%&gt;%</span></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">total_cases =</span> <span class="fu">sum</span>(cases), <span class="at">.by =</span> country) <span class="sc">%&gt;%</span></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(total_cases)) <span class="sc">%&gt;%</span></span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>top_10_countries</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 2
   country                               total_cases
   &lt;chr&gt;                                       &lt;dbl&gt;
 1 China                                      869092
 2 India                                      630164
 3 South Africa                               364992
 4 Indonesia                                  296272
 5 Bangladesh                                 150856
 6 Pakistan                                   124455
 7 Russian Federation                         102823
 8 Philippines                                 89198
 9 Kenya                                       85484
10 Democratic People's Republic of Korea       81240</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>who_tidy <span class="sc">%&gt;%</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>    country <span class="sc">%in%</span> top_10_countries<span class="sc">$</span>country,</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>    year <span class="sc">&gt;=</span> <span class="dv">2000</span>,</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>    year <span class="sc">&lt;=</span> <span class="dv">2010</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">total_cases =</span> <span class="fu">sum</span>(cases), <span class="at">.by =</span> <span class="fu">c</span>(country, year)</span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> year, <span class="at">y =</span> total_cases, <span class="at">col =</span> country)) <span class="sc">+</span></span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="chapter_5_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid figure-img" width="2000"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="summary" class="level2" data-number="5.8">
<h2 data-number="5.8" class="anchored" data-anchor-id="summary"><span class="header-section-number">5.8</span> Summary</h2>
<p>Today we learned about the concept of “tidy data”. A particular form of tabular data that is particularly suitable for the tidyverse. It contains:</p>
<ul>
<li>one variable per column</li>
<li>one observation per row</li>
<li>one value per cell</li>
</ul>
<p>We learned to reshape data frames using various functions powerful functions.</p>
<ul>
<li><code>pivot_wider()</code> reduces the number of rows and move information into new columns.</li>
<li><code>pivot_longer()</code> reduces the number of columns and increase the number of rows. Moving the column names into a new column.</li>
<li><code>separate()</code> separates multiple values encoded in a single column into individual columns based on a separator character.</li>
</ul>
<p>We learned to combine data frames using the <code>left_join()</code> function. This allows us to merge data frames together that have common columns, matching up the rows according to the shared column.</p>
<ul>
<li><code>left_join(left, right, by = shared_column)</code> takes two data frames and merges them based on the <code>shared_column</code> column. The information from the right data frame is added to the left if there is a match within the shared column.</li>
<li><code>left_join(left, right, by = join_by(left_col_name == right_col_name))</code> can be used to join using columns from two data frames that encode the same values but have different column names</li>
</ul>
<p>We also learned some useful string manipulation functions that help us clean up data columns</p>
<ul>
<li><code>str_replace()</code> lets us replace patterns in a string</li>
<li><code>str_remove()</code> lets us remove patterns in a string</li>
</ul>
<p>Together these make up a very versatile toolkit for transforming messy data you find in the wild into a tidy, more useful format to summarise information or produce plots with.</p>
<!-- :::::::::::::::::::: {.callout-important title="Practice exercises"} -->
<!-- Try these practice questions to test your understanding -->
<!-- :::::::: question -->
<!-- 1\. What does the following code do? -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- m_dose %>% -->
<!--   summarise( -->
<!--     med_tail = median(tail_length_mm, na.rm = TRUE), -->
<!--     .by = c(mouse_strain, sex)) %>% -->
<!--   pivot_wider(names_from = sex, values_from = med_tail) -->
<!-- ``` -->
<!-- ::::::: choices -->
<!-- ::: choice -->
<!-- Pivots data into a wide format where there is a column for each sex. -->
<!-- ::: -->
<!-- ::: {.choice .correct-choice} -->
<!-- Calculates the median tail length for each unique combination of `mouse_strain` and `sex` in the `m_dose` data frame, then pivots into a wide format where there is a column for each sex. -->
<!-- ::: -->
<!-- ::: choice -->
<!-- Calculates the median tail length for each unique combination of `mouse_strain` and `sex` in the `m_dose` data frame, then pivots into a wide format where there is a column for each mouse strain. -->
<!-- ::: -->
<!-- ::: choice -->
<!-- It just gives an error -->
<!-- ::: -->
<!-- ::::::: -->
<!-- :::::::: -->
<!-- 2\. I have run the following code to create a new column in the `m_dose` data frame that gives the weight of the mice at the end of the experiment. -->
<!-- ```{r} -->
<!-- m_dose %>% -->
<!--   # add a column for the weight at the end of the experiment -->
<!--   mutate(final_weight_g = initial_weight_g - weight_lost_g) %>% -->
<!--   # select the relevant columns only -->
<!--   select(id_num, initial_weight_g, final_weight_g) -->
<!-- ``` -->
<!-- Which pivot function call would I use to take this data from a wide format (where there is a column for the final and initial weight) to a long format (where there is a row for each mouse and each weight measurement)? -->
<!-- ::::::: choices -->
<!-- ::: choice -->
<!-- `pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "weight", values_to = "timepoint")` -->
<!-- ::: -->
<!-- ::: choice -->
<!-- `pivot_longer(cols = c(id_num, final_weight_g), names_to = "timepoint", values_to = "initial_weight_g")` -->
<!-- ::: -->
<!-- ::: choice -->
<!-- `pivot_wider(names_from = initial_weight_g, values_from = final_weight_g)` -->
<!-- ::: -->
<!-- ::: {.choice .correct-choice} -->
<!-- `pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")` -->
<!-- ::: -->
<!-- ::::::: -->
<!-- :::::::: question -->
<!-- 3\. Which of the following is NOT a valid way to join the `m_dose` data frame with the `m_exp` data frame based on the `id_num` column? -->
<!-- ::::::: choices -->
<!-- ::: choice -->
<!-- `m_dose %>% left_join(m_exp, by = "id_num")` -->
<!-- ::: -->
<!-- ::: choice -->
<!-- `left_join(m_dose, m_exp, by = "id_num")` -->
<!-- ::: -->
<!-- ::: {.choice .correct-choice} -->
<!-- `m_dose %>% left_join(m_exp, .by = "id_num")` -->
<!-- ::: -->
<!-- ::: choice -->
<!-- `m_dose %>% left_join(m_exp, by = ("id_num" = "id_num"))` -->
<!-- ::: -->
<!-- ::::::: -->
<!-- :::::::: -->
<!-- <details> -->
<!-- <summary>Solutions</summary> -->
<!-- 1\. The code first calculates the median tail length for each unique combination of `mouse_strain` and `sex` in the `m_dose` data frame, then pivots the data into a wide format where there is a column for each sex in the dataset (because of the argument `names_from = sex` ) -->
<!-- 2\. The correct pivot function call to take the data from a wide format to a long format is `pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")`. This code tells R to pivot the `initial_weight_g` and `final_weight_g` columns into a long format, where there is a row for each mouse and each weight measurement. The `names_to` argument specifies to make a column called 'timepoint' that tells us whether the measurement is initial or final, and the `values_to` argument specifies the name of the new column that will contain these measurements. -->
<!-- 3\. The line of code that is NOT a valid way to join the `m_dose` data frame with the `m_exp` data frame based on the `id_num` column is `m_dose %>% left_join(m_exp, .by = "id_num")`. This line of code is incorrect because the `.by` argument is not used in the `left_join()` function (this can be confusing! it's `.by` when grouping by `by` when joining). The other options are valid ways to join the two data frames based on the `id_num` column: remember that we don't have to use pipes to join data frames, we can use the `left_join()` function directly, and we can use a named vector to specify the columns to join on (although here it's a bit redundant as the columns have the same name). -->
<!-- </details> -->
<!-- :::::::::::::::::::: -->
<!-- summary - Reshaping data with `pivot_longer()` and `pivot_wider()` to change the structure of your data frame. -->
<!-- -   Combining data with `left_join()` to merge two data frames based on a common column. -->
<!-- questions etc -->
<!-- 4.  Using the `m_dose` data frame, write R code to: -->
<!--     a.  Make a data frame that shows the number of mice of each strain, in each replicate. -->
<!--     b.  Pivot this data frame into a wide format to create a contingency table. -->
<!--     c.  Pivot the wide data frame from (b) back into a long format. -->
<!-- 5.  Let's say I have two data frames, `df1` and `df2`, that I want to join based a shared 'key' column, that is called 'key' in `df1` and 'item_key' in `df2`. Write R code to join these two data frames using the `left_join()` function. -->
<!-- solutions -->
<!-- 4.  Here's how you could write R code to achieve the tasks: -->
<!--     a.  `m_dose %>% summarise(n_mice = n(), .by = c(mouse_strain, replicate))` -->
<!--     b.  `m_dose %>% summarise(n_mice = n(), .by = c(mouse_strain, replicate)) %>% pivot_wider(names_from = replicate, values_from = n_mice)` -->
<!--     c.  `m_dose %>% summarise(n_mice = n(), .by = c(mouse_strain, replicate)) %>% pivot_wider(names_from = replicate, values_from = n_mice) %>% pivot_longer(cols = starts_with("rep"), names_to = "replicate", values_to = "n_mice")` -->
<!-- 5.  To join the two data frames you could use `df1 %>% left_join(df2, by = c("key" = "item_key"))` (with pipe) or `left_join(df1, df2, by = c("key" = "item_key"))` (without pipe). -->


<!-- -->

</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./chapter_4.html" class="pagination-link" aria-label="Plotting with ggplot2">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Plotting with ggplot2</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./stats_primer.html" class="pagination-link" aria-label="Statistics in R Primer">
        <span class="nav-page-text">Statistics in R Primer</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb68" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="an">filters:</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a><span class="co">  - naquiz</span></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-location: left</span></span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a><span class="co">    toc-title: "In this chapter:"</span></span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a><span class="fu"># Putting it all together {#sec-chapter05}</span></span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a>In this chapter we will combine all the skills we have learned so far to perform a complete analysis of a small dataset.</span>
<span id="cb68-14"><a href="#cb68-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-15"><a href="#cb68-15" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip title="Learning Objectives"}</span>
<span id="cb68-16"><a href="#cb68-16" aria-hidden="true" tabindex="-1"></a>At the end of this chapter, learners should be able to:</span>
<span id="cb68-17"><a href="#cb68-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-18"><a href="#cb68-18" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Describe the key steps in data analysis (exploration, manipulating and plotting)</span>
<span id="cb68-19"><a href="#cb68-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-20"><a href="#cb68-20" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Understand how pivot and join functions can be used to reshape and combine entire data frames</span>
<span id="cb68-21"><a href="#cb68-21" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb68-22"><a href="#cb68-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-23"><a href="#cb68-23" aria-hidden="true" tabindex="-1"></a><span class="fu">## Introduction to the dataset</span></span>
<span id="cb68-24"><a href="#cb68-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-25"><a href="#cb68-25" aria-hidden="true" tabindex="-1"></a>In this chapter we will use the mousezempic dosage data <span class="in">`m_dose`</span> and mouse expression data <span class="in">`m_exp`</span> data frames, which contain information about the mice including mousezempic dose, and their gene expression levels, respectively.</span>
<span id="cb68-26"><a href="#cb68-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-29"><a href="#cb68-29" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-30"><a href="#cb68-30" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb68-31"><a href="#cb68-31" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb68-32"><a href="#cb68-32" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"data"</span>) <span class="co"># create data directory if it doesn't exist</span></span>
<span id="cb68-33"><a href="#cb68-33" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://raw.githubusercontent.com/kzeglinski/new_wehi_r_course/refs/heads/main/data/m_exp.tsv"</span>, <span class="st">"data/m_exp.tsv"</span>) <span class="co"># download data</span></span>
<span id="cb68-34"><a href="#cb68-34" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-35"><a href="#cb68-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-38"><a href="#cb68-38" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-39"><a href="#cb68-39" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb68-40"><a href="#cb68-40" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb68-41"><a href="#cb68-41" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb68-42"><a href="#cb68-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-43"><a href="#cb68-43" aria-hidden="true" tabindex="-1"></a><span class="co"># read in dosage data</span></span>
<span id="cb68-44"><a href="#cb68-44" aria-hidden="true" tabindex="-1"></a>m_dose <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/mousezempic_dosage_data.csv"</span>)</span>
<span id="cb68-45"><a href="#cb68-45" aria-hidden="true" tabindex="-1"></a>m_dose</span>
<span id="cb68-46"><a href="#cb68-46" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-47"><a href="#cb68-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-50"><a href="#cb68-50" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-51"><a href="#cb68-51" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb68-52"><a href="#cb68-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-53"><a href="#cb68-53" aria-hidden="true" tabindex="-1"></a><span class="co"># read in expression data</span></span>
<span id="cb68-54"><a href="#cb68-54" aria-hidden="true" tabindex="-1"></a>m_exp <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"data/mousezempic_expression_data.tsv"</span>)</span>
<span id="cb68-55"><a href="#cb68-55" aria-hidden="true" tabindex="-1"></a>m_exp</span>
<span id="cb68-56"><a href="#cb68-56" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-57"><a href="#cb68-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-58"><a href="#cb68-58" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb68-59"><a href="#cb68-59" aria-hidden="true" tabindex="-1"></a>We have used <span class="in">`read_csv()`</span> for the mouse data and <span class="in">`read_tsv()`</span> for the expression data.</span>
<span id="cb68-60"><a href="#cb68-60" aria-hidden="true" tabindex="-1"></a>These are for reading data separated by commas and tab characters respectively.</span>
<span id="cb68-61"><a href="#cb68-61" aria-hidden="true" tabindex="-1"></a>The <span class="in">`readr`</span> package also provides <span class="in">`read_delim()`</span> to let the package guess your delimiter, but if you know the format of your file then it's good practice to use the appropriate reading function for more predictable behaviour.</span>
<span id="cb68-62"><a href="#cb68-62" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb68-63"><a href="#cb68-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-64"><a href="#cb68-64" aria-hidden="true" tabindex="-1"></a>::: callout-caution</span>
<span id="cb68-65"><a href="#cb68-65" aria-hidden="true" tabindex="-1"></a>There's nothing stopping someone from naming a file <span class="in">`file.csv`</span> while having tab-separated data inside.</span>
<span id="cb68-66"><a href="#cb68-66" aria-hidden="true" tabindex="-1"></a>This happens quite often in real-world data so it's a good idea to have a quick look at the data in an text editor before reading it in.</span>
<span id="cb68-67"><a href="#cb68-67" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb68-68"><a href="#cb68-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-69"><a href="#cb68-69" aria-hidden="true" tabindex="-1"></a><span class="fu">## Tidy Data</span></span>
<span id="cb68-70"><a href="#cb68-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-71"><a href="#cb68-71" aria-hidden="true" tabindex="-1"></a>The tidyverse revolves around an important concept called "<span class="co">[</span><span class="ot">tidy data</span><span class="co">](https://r4ds.hadley.nz/data-tidy.html)</span>".</span>
<span id="cb68-72"><a href="#cb68-72" aria-hidden="true" tabindex="-1"></a>This is a specific representation of tabular data that is considered easy to work with.</span>
<span id="cb68-73"><a href="#cb68-73" aria-hidden="true" tabindex="-1"></a>Tidy data is roughly defined as tabular data that contains:</span>
<span id="cb68-74"><a href="#cb68-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-75"><a href="#cb68-75" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Individual variables in columns</span>
<span id="cb68-76"><a href="#cb68-76" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Individual observations in rows</span>
<span id="cb68-77"><a href="#cb68-77" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>One data value in each cell</span>
<span id="cb68-78"><a href="#cb68-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-79"><a href="#cb68-79" aria-hidden="true" tabindex="-1"></a>Having individual variables in columns makes then accessible for performing tidyverse operations like <span class="in">`select()`</span>, <span class="in">`mutate()`</span>, <span class="in">`filter()`</span> and <span class="in">`arrange()`</span>.</span>
<span id="cb68-80"><a href="#cb68-80" aria-hidden="true" tabindex="-1"></a>If variables were not stored as columns, then these functions would not be able to access them by name.</span>
<span id="cb68-81"><a href="#cb68-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-82"><a href="#cb68-82" aria-hidden="true" tabindex="-1"></a>Having individual observations in rows is important because it associates all variables of each observation with the same row.</span>
<span id="cb68-83"><a href="#cb68-83" aria-hidden="true" tabindex="-1"></a>If the data from one observation is spread across multiple rows then it is easy to incorrect summaries from the <span class="in">`summarise()`</span> function.</span>
<span id="cb68-84"><a href="#cb68-84" aria-hidden="true" tabindex="-1"></a>When using the <span class="in">`filter()`</span> function with tidy data, you can expect to keep all the data for an observation or none at all.</span>
<span id="cb68-85"><a href="#cb68-85" aria-hidden="true" tabindex="-1"></a>When the data for observations is split over different rows, it's possible to unknowingly lose partial data from observations.</span>
<span id="cb68-86"><a href="#cb68-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-87"><a href="#cb68-87" aria-hidden="true" tabindex="-1"></a>Having a single value in each cell makes it possible to perform meaningful computations for the values, for example you cannot take a <span class="in">`mean()`</span> of a column of values that contain multiple different values.</span>
<span id="cb68-88"><a href="#cb68-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-89"><a href="#cb68-89" aria-hidden="true" tabindex="-1"></a>Although tidy data is the easiest to work with, it's often necessary to alter the format of your data for plotting or table displays.</span>
<span id="cb68-90"><a href="#cb68-90" aria-hidden="true" tabindex="-1"></a>It's a good idea to keep your core data in a tidy format and treating plot or table outputs as representations of that tidy data.</span>
<span id="cb68-91"><a href="#cb68-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-92"><a href="#cb68-92" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb68-93"><a href="#cb68-93" aria-hidden="true" tabindex="-1"></a>Most data you encounter will not be tidy, the first part of data analysis is usually called "data-wranging" and involves tidying up your data so it is easier to use for downstream analysis.</span>
<span id="cb68-94"><a href="#cb68-94" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb68-95"><a href="#cb68-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-96"><a href="#cb68-96" aria-hidden="true" tabindex="-1"></a><span class="fu">## Reshaping and combining data {#sec-reshaping}</span></span>
<span id="cb68-97"><a href="#cb68-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-98"><a href="#cb68-98" aria-hidden="true" tabindex="-1"></a>The <span class="in">`filter()`</span>, <span class="in">`select()`</span>, <span class="in">`mutate()`</span> and <span class="in">`summarise()`</span> functions we learned last chapter all operate along either the columns or the rows of data.</span>
<span id="cb68-99"><a href="#cb68-99" aria-hidden="true" tabindex="-1"></a>Combining these operations cleverly can answer the majority of questions about your data.</span>
<span id="cb68-100"><a href="#cb68-100" aria-hidden="true" tabindex="-1"></a>However, there are two useful families of functions: <span class="in">`pivot`</span> for reshaping your data and <span class="in">`join`</span> for combining your data from shared columns.</span>
<span id="cb68-101"><a href="#cb68-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-102"><a href="#cb68-102" aria-hidden="true" tabindex="-1"></a><span class="fu">### Reshaping data with pivot functions {#sec-pivot}</span></span>
<span id="cb68-103"><a href="#cb68-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-104"><a href="#cb68-104" aria-hidden="true" tabindex="-1"></a>Pivoting is a way to change the shape of your tibble.</span>
<span id="cb68-105"><a href="#cb68-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-106"><a href="#cb68-106" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Pivoting longer reshapes the data to transfer data stored in columns into rows, resulting in more rows and fewer columns.</span>
<span id="cb68-107"><a href="#cb68-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-108"><a href="#cb68-108" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Pivoting wider is the reverse, moving data from rows into columns.</span>
<span id="cb68-109"><a href="#cb68-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-110"><a href="#cb68-110" aria-hidden="true" tabindex="-1"></a><span class="al">![Pivot functions allow you to change the structure of your data frame](images/S02F10_pivot.png)</span>{width="50%"}</span>
<span id="cb68-111"><a href="#cb68-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-112"><a href="#cb68-112" aria-hidden="true" tabindex="-1"></a>The <span class="in">`pivot_longer()`</span> function is used to pivot data from wide to long format, and the <span class="in">`pivot_wider()`</span> function is used to pivot data from long to wide format.</span>
<span id="cb68-113"><a href="#cb68-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-114"><a href="#cb68-114" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Pivot wider</span></span>
<span id="cb68-115"><a href="#cb68-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-116"><a href="#cb68-116" aria-hidden="true" tabindex="-1"></a>A common use case for <span class="in">`pivot_wider()`</span> is to make a <span class="co">[</span><span class="ot">contingency table</span><span class="co">](https://en.wikipedia.org/wiki/Contingency_table)</span>, which shows the number of observations for each combination of two variables.</span>
<span id="cb68-117"><a href="#cb68-117" aria-hidden="true" tabindex="-1"></a>This is often easier to read than the same information in long format.</span>
<span id="cb68-118"><a href="#cb68-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-119"><a href="#cb68-119" aria-hidden="true" tabindex="-1"></a>For example, let's say we want to create a table that shows how many mice there are of each strain, in each cage number.</span>
<span id="cb68-120"><a href="#cb68-120" aria-hidden="true" tabindex="-1"></a>We can achieve this in a long format using <span class="in">`summarise()`</span> as we learned in the previous section:</span>
<span id="cb68-121"><a href="#cb68-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-124"><a href="#cb68-124" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-125"><a href="#cb68-125" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb68-126"><a href="#cb68-126" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb68-127"><a href="#cb68-127" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_mice =</span> <span class="fu">n</span>(),</span>
<span id="cb68-128"><a href="#cb68-128" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(cage_number, mouse_strain))</span>
<span id="cb68-129"><a href="#cb68-129" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-130"><a href="#cb68-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-131"><a href="#cb68-131" aria-hidden="true" tabindex="-1"></a>For the specific task of counting, we can achieve the same effect using the <span class="in">`count()`</span> tidyverse function.</span>
<span id="cb68-132"><a href="#cb68-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-135"><a href="#cb68-135" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-136"><a href="#cb68-136" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb68-137"><a href="#cb68-137" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(cage_number, mouse_strain, <span class="at">name =</span> <span class="st">"n_mice"</span>)</span>
<span id="cb68-138"><a href="#cb68-138" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-139"><a href="#cb68-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-140"><a href="#cb68-140" aria-hidden="true" tabindex="-1"></a>This summarises by each combination of <span class="in">`cage_number`</span> and <span class="in">`mouse_strain`</span>, with the <span class="in">`n()`</span> function giving the count of data belonging to that combination.</span>
<span id="cb68-141"><a href="#cb68-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-142"><a href="#cb68-142" aria-hidden="true" tabindex="-1"></a>To get a contingency table, we wish to have the information in the <span class="in">`mouse_strain`</span> column displayed along the column names and the values of <span class="in">`n_mice`</span> becoming the values of the cells in the new table.</span>
<span id="cb68-143"><a href="#cb68-143" aria-hidden="true" tabindex="-1"></a>Since the goal is to make the table wider, we use the <span class="in">`pivot_wider()`</span> function.</span>
<span id="cb68-144"><a href="#cb68-144" aria-hidden="true" tabindex="-1"></a>To achieve this, we instruct the <span class="in">`pivot_wider()`</span> function to take names from the <span class="in">`mouse_strain`</span> column and the values from the <span class="in">`n_mice`</span> column.</span>
<span id="cb68-145"><a href="#cb68-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-148"><a href="#cb68-148" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-149"><a href="#cb68-149" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb68-150"><a href="#cb68-150" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(cage_number, mouse_strain, <span class="at">name =</span> <span class="st">"n_mice"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb68-151"><a href="#cb68-151" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span>
<span id="cb68-152"><a href="#cb68-152" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-153"><a href="#cb68-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-154"><a href="#cb68-154" aria-hidden="true" tabindex="-1"></a>This has transformed our data into a contingency table, with <span class="in">`NA`</span> where no data corresponding data exists for the specific <span class="in">`cage_number`</span> and <span class="in">`mouse_strain`</span> combination.</span>
<span id="cb68-155"><a href="#cb68-155" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-156"><a href="#cb68-156" aria-hidden="true" tabindex="-1"></a>We can do the same thing to see how many of each mouse strains is in each of our experiment replicates.</span>
<span id="cb68-157"><a href="#cb68-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-160"><a href="#cb68-160" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-161"><a href="#cb68-161" aria-hidden="true" tabindex="-1"></a>m_dose <span class="sc">%&gt;%</span></span>
<span id="cb68-162"><a href="#cb68-162" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(replicate, mouse_strain, <span class="at">name =</span> <span class="st">"n_mice"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb68-163"><a href="#cb68-163" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> mouse_strain, <span class="at">values_from =</span> n_mice)</span>
<span id="cb68-164"><a href="#cb68-164" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-165"><a href="#cb68-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-166"><a href="#cb68-166" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Pivot longer</span></span>
<span id="cb68-167"><a href="#cb68-167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-168"><a href="#cb68-168" aria-hidden="true" tabindex="-1"></a>Data can often arrive in the form similar to the contingency table we constructed.</span>
<span id="cb68-169"><a href="#cb68-169" aria-hidden="true" tabindex="-1"></a>Although this data is easy to read, it is difficult to operate on using tidyverse functions because the <span class="in">`mouse_strain`</span> data is now stored in the column names and not inside a column we can use as a variable.</span>
<span id="cb68-170"><a href="#cb68-170" aria-hidden="true" tabindex="-1"></a>In order to make this kind of data tidy, we use the <span class="in">`pivot_longer()`</span> function, which will create a pair of columns from the column names and the value of the corresponding cell.</span>
<span id="cb68-171"><a href="#cb68-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-172"><a href="#cb68-172" aria-hidden="true" tabindex="-1"></a>To demonstrate <span class="in">`pivot_longer()`</span>, we will use the <span class="in">`m_exp`</span> that we downloaded earlier.</span>
<span id="cb68-173"><a href="#cb68-173" aria-hidden="true" tabindex="-1"></a>This data frame contains the expression levels of two genes (TH and PRLH) suspected to be upregulated in mice taking MouseZempic, as well as one housekeeping gene (HPRT1), all measured in triplicate.</span>
<span id="cb68-174"><a href="#cb68-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-177"><a href="#cb68-177" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-178"><a href="#cb68-178" aria-hidden="true" tabindex="-1"></a>m_exp</span>
<span id="cb68-179"><a href="#cb68-179" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-180"><a href="#cb68-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-181"><a href="#cb68-181" aria-hidden="true" tabindex="-1"></a>The data is currently in wide format, with each row representing a different mouse (identified by its <span class="in">`id_num`</span>) and each column representing a different measurement of a gene.</span>
<span id="cb68-182"><a href="#cb68-182" aria-hidden="true" tabindex="-1"></a>To reshape this data into a long format (where each measurement is contained on a separate row), we can use <span class="in">`pivot_longer()`</span>, specifying three arguments:</span>
<span id="cb68-183"><a href="#cb68-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-184"><a href="#cb68-184" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span><span class="in">`cols`</span>: the columns to pivot from. You can use selection helpers like <span class="in">`contains()`</span> or <span class="in">`starts_with()`</span> to easily select multiple columns at once.</span>
<span id="cb68-185"><a href="#cb68-185" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span><span class="in">`names_to`</span>: the name of a new column that will contain the original column names.</span>
<span id="cb68-186"><a href="#cb68-186" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span><span class="in">`values_to`</span>: the name of a new column that will contain the values from the original columns.</span>
<span id="cb68-187"><a href="#cb68-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-188"><a href="#cb68-188" aria-hidden="true" tabindex="-1"></a>In this particular case here's what the code would look like:</span>
<span id="cb68-189"><a href="#cb68-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-192"><a href="#cb68-192" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-193"><a href="#cb68-193" aria-hidden="true" tabindex="-1"></a>m_exp_long <span class="ot">&lt;-</span> m_exp <span class="sc">%&gt;%</span></span>
<span id="cb68-194"><a href="#cb68-194" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb68-195"><a href="#cb68-195" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="fu">contains</span>(<span class="st">"_rep"</span>),</span>
<span id="cb68-196"><a href="#cb68-196" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"measurement"</span>,</span>
<span id="cb68-197"><a href="#cb68-197" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">"expression_level"</span></span>
<span id="cb68-198"><a href="#cb68-198" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb68-199"><a href="#cb68-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-200"><a href="#cb68-200" aria-hidden="true" tabindex="-1"></a>m_exp_long</span>
<span id="cb68-201"><a href="#cb68-201" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-202"><a href="#cb68-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-203"><a href="#cb68-203" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Perplexed by pivoting?"}</span>
<span id="cb68-204"><a href="#cb68-204" aria-hidden="true" tabindex="-1"></a>Pivoting can be a bit tricky to get your head around!</span>
<span id="cb68-205"><a href="#cb68-205" aria-hidden="true" tabindex="-1"></a>Often when you're doing analysis, you'll run into the problem of knowing that you need to pivot, but not knowing exactly what arguments to use.</span>
<span id="cb68-206"><a href="#cb68-206" aria-hidden="true" tabindex="-1"></a>In these cases, it can be helpful to look at examples online, <span class="co">[</span><span class="ot">like those in the R for Data Science book</span><span class="co">](https://r4ds.hadley.nz/data-tidy.html#sec-pivoting)</span>.</span>
<span id="cb68-207"><a href="#cb68-207" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb68-208"><a href="#cb68-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-209"><a href="#cb68-209" aria-hidden="true" tabindex="-1"></a><span class="fu">### Separating data in a column</span></span>
<span id="cb68-210"><a href="#cb68-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-211"><a href="#cb68-211" aria-hidden="true" tabindex="-1"></a>When we look at the <span class="in">`measurement`</span> column we see that it contains two pieces of information.</span>
<span id="cb68-212"><a href="#cb68-212" aria-hidden="true" tabindex="-1"></a>The gene being measured and the replicate number separated by "<span class="sc">\_</span>".</span>
<span id="cb68-213"><a href="#cb68-213" aria-hidden="true" tabindex="-1"></a>We can use the <span class="in">`separate()`</span> function to split this data into individual columns so that one column does not contain multiple variables.</span>
<span id="cb68-214"><a href="#cb68-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-217"><a href="#cb68-217" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-218"><a href="#cb68-218" aria-hidden="true" tabindex="-1"></a>m_exp_separate <span class="ot">&lt;-</span> m_exp_long <span class="sc">%&gt;%</span></span>
<span id="cb68-219"><a href="#cb68-219" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(measurement, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"gene"</span>, <span class="st">"replicate"</span>), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb68-220"><a href="#cb68-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-221"><a href="#cb68-221" aria-hidden="true" tabindex="-1"></a>m_exp_separate</span>
<span id="cb68-222"><a href="#cb68-222" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-223"><a href="#cb68-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-224"><a href="#cb68-224" aria-hidden="true" tabindex="-1"></a><span class="fu">### Summarising values</span></span>
<span id="cb68-225"><a href="#cb68-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-226"><a href="#cb68-226" aria-hidden="true" tabindex="-1"></a>Now suppose we wanted to combine the gene expression data across replicates by adding them up, we can use the <span class="in">`summarise()`</span> function to do so.</span>
<span id="cb68-227"><a href="#cb68-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-230"><a href="#cb68-230" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-231"><a href="#cb68-231" aria-hidden="true" tabindex="-1"></a>m_exp_rep_summed <span class="ot">&lt;-</span> m_exp_separate <span class="sc">%&gt;%</span></span>
<span id="cb68-232"><a href="#cb68-232" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb68-233"><a href="#cb68-233" aria-hidden="true" tabindex="-1"></a>    <span class="at">expression_level =</span> <span class="fu">sum</span>(expression_level),</span>
<span id="cb68-234"><a href="#cb68-234" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(id_num, group, gene)</span>
<span id="cb68-235"><a href="#cb68-235" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb68-236"><a href="#cb68-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-237"><a href="#cb68-237" aria-hidden="true" tabindex="-1"></a>m_exp_rep_summed</span>
<span id="cb68-238"><a href="#cb68-238" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-239"><a href="#cb68-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-240"><a href="#cb68-240" aria-hidden="true" tabindex="-1"></a>We may also want to get the average expression level of each gene within each experimental group and plot it as a bar chart.</span>
<span id="cb68-241"><a href="#cb68-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-244"><a href="#cb68-244" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-245"><a href="#cb68-245" aria-hidden="true" tabindex="-1"></a>m_exp_rep_summed <span class="sc">%&gt;%</span></span>
<span id="cb68-246"><a href="#cb68-246" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb68-247"><a href="#cb68-247" aria-hidden="true" tabindex="-1"></a>    <span class="at">expression_level =</span> <span class="fu">mean</span>(expression_level),</span>
<span id="cb68-248"><a href="#cb68-248" aria-hidden="true" tabindex="-1"></a>    <span class="at">.by =</span> <span class="fu">c</span>(group, gene)</span>
<span id="cb68-249"><a href="#cb68-249" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb68-250"><a href="#cb68-250" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> gene, <span class="at">y =</span> expression_level, <span class="at">fill =</span> group)) <span class="sc">+</span></span>
<span id="cb68-251"><a href="#cb68-251" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>(<span class="at">position =</span> <span class="st">"dodge"</span>)</span>
<span id="cb68-252"><a href="#cb68-252" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-253"><a href="#cb68-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-254"><a href="#cb68-254" aria-hidden="true" tabindex="-1"></a>We see that all 3 genes are slightly higher in the treatment group.</span>
<span id="cb68-255"><a href="#cb68-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-256"><a href="#cb68-256" aria-hidden="true" tabindex="-1"></a><span class="fu">### Reshaping for plotting</span></span>
<span id="cb68-257"><a href="#cb68-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-258"><a href="#cb68-258" aria-hidden="true" tabindex="-1"></a>Suppose we wanted to plot the value of two genes against each other.</span>
<span id="cb68-259"><a href="#cb68-259" aria-hidden="true" tabindex="-1"></a>We would need the expression values of genes to be in individual columns.</span>
<span id="cb68-260"><a href="#cb68-260" aria-hidden="true" tabindex="-1"></a>However we have lost this structure after using <span class="in">`pivot_longer()`</span>.</span>
<span id="cb68-261"><a href="#cb68-261" aria-hidden="true" tabindex="-1"></a>We can restore it using <span class="in">`pivot_wider()`</span> after we have summarised the replicates.</span>
<span id="cb68-262"><a href="#cb68-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-265"><a href="#cb68-265" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-266"><a href="#cb68-266" aria-hidden="true" tabindex="-1"></a>m_exp_rep_summed_wide <span class="ot">&lt;-</span> m_exp_rep_summed <span class="sc">%&gt;%</span> </span>
<span id="cb68-267"><a href="#cb68-267" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>( <span class="at">names_from =</span> <span class="st">"gene"</span>, <span class="at">values_from =</span> <span class="st">"expression_level"</span>)</span>
<span id="cb68-268"><a href="#cb68-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-269"><a href="#cb68-269" aria-hidden="true" tabindex="-1"></a>m_exp_rep_summed_wide</span>
<span id="cb68-270"><a href="#cb68-270" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-271"><a href="#cb68-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-272"><a href="#cb68-272" aria-hidden="true" tabindex="-1"></a>With the replicate aggregated expression levels in individual columns, we can now plot the gene expression values against each other.</span>
<span id="cb68-273"><a href="#cb68-273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-276"><a href="#cb68-276" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-277"><a href="#cb68-277" aria-hidden="true" tabindex="-1"></a>m_exp_rep_summed_wide <span class="sc">%&gt;%</span> <span class="fu">ggplot</span>( <span class="fu">aes</span>(<span class="at">x =</span> Th, <span class="at">y =</span> Prlh)) <span class="sc">+</span></span>
<span id="cb68-278"><a href="#cb68-278" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span>
<span id="cb68-279"><a href="#cb68-279" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-280"><a href="#cb68-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-281"><a href="#cb68-281" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Practice exercise"}</span>
<span id="cb68-282"><a href="#cb68-282" aria-hidden="true" tabindex="-1"></a>Make all the pair-wise scatter plots and assemble them together using <span class="in">`patchwork`</span>.</span>
<span id="cb68-283"><a href="#cb68-283" aria-hidden="true" tabindex="-1"></a>There should be plots of:</span>
<span id="cb68-284"><a href="#cb68-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-285"><a href="#cb68-285" aria-hidden="true" tabindex="-1"></a>Th vs Prlh</span>
<span id="cb68-286"><a href="#cb68-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-287"><a href="#cb68-287" aria-hidden="true" tabindex="-1"></a>Th vs Hprt1</span>
<span id="cb68-288"><a href="#cb68-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-289"><a href="#cb68-289" aria-hidden="true" tabindex="-1"></a>Prlh vs Hprt1</span>
<span id="cb68-290"><a href="#cb68-290" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb68-291"><a href="#cb68-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-292"><a href="#cb68-292" aria-hidden="true" tabindex="-1"></a><span class="fu">### Combining data with join functions {#sec-join}</span></span>
<span id="cb68-293"><a href="#cb68-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-294"><a href="#cb68-294" aria-hidden="true" tabindex="-1"></a>Data analysis typically involves multiple tables of data.</span>
<span id="cb68-295"><a href="#cb68-295" aria-hidden="true" tabindex="-1"></a>Often there will be tables that are contain related information that must be combined to answer the questions you're interested in.</span>
<span id="cb68-296"><a href="#cb68-296" aria-hidden="true" tabindex="-1"></a>Tables that are related to each other tend to have one or more columns in common, and are referred to as "relational data".</span>
<span id="cb68-297"><a href="#cb68-297" aria-hidden="true" tabindex="-1"></a>Combining relational data in useful ways requires the <span class="in">`join`</span> family of functions.</span>
<span id="cb68-298"><a href="#cb68-298" aria-hidden="true" tabindex="-1"></a>In general joins can accomplish two tasks</span>
<span id="cb68-299"><a href="#cb68-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-300"><a href="#cb68-300" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Add new variables to an existing table containing additional information.</span>
<span id="cb68-301"><a href="#cb68-301" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Filter observations in one table based on whether or not they match observations in another table.</span>
<span id="cb68-302"><a href="#cb68-302" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Both of the above at the same time.</span>
<span id="cb68-303"><a href="#cb68-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-304"><a href="#cb68-304" aria-hidden="true" tabindex="-1"></a><span class="al">![The left_join() function allows you to combine two data frames](images/S02F11_left_join.png)</span></span>
<span id="cb68-305"><a href="#cb68-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-306"><a href="#cb68-306" aria-hidden="true" tabindex="-1"></a>A common and basic join is the <span class="in">`left_join()`</span>.</span>
<span id="cb68-307"><a href="#cb68-307" aria-hidden="true" tabindex="-1"></a>It takes two data frame as arguments and optionally a vector of common columns to perform the join on.</span>
<span id="cb68-308"><a href="#cb68-308" aria-hidden="true" tabindex="-1"></a>The reason it's called a left-join because it retains all rows from the left data frame while adding on columns from the right data frame only when the data in the designed joining column(s) match.</span>
<span id="cb68-309"><a href="#cb68-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-310"><a href="#cb68-310" aria-hidden="true" tabindex="-1"></a>For example, we can join the <span class="in">`m_dose`</span> data frame with the <span class="in">`m_exp_rep_summed_wide`</span> data frame based on the <span class="in">`id_num`</span> column, which is common to both data frames:</span>
<span id="cb68-311"><a href="#cb68-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-314"><a href="#cb68-314" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-315"><a href="#cb68-315" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>(m_dose, m_exp_rep_summed_wide, <span class="at">by =</span> <span class="st">"id_num"</span>)</span>
<span id="cb68-316"><a href="#cb68-316" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-317"><a href="#cb68-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-318"><a href="#cb68-318" aria-hidden="true" tabindex="-1"></a>The <span class="in">`m_dose`</span> contains information about characteristics of each mouse, while the <span class="in">`m_exp_rep_summed_wide`</span> contains the replicate-summed gene expressions of the mice.</span>
<span id="cb68-319"><a href="#cb68-319" aria-hidden="true" tabindex="-1"></a>Each data frame has an <span class="in">`id_num`</span> column that relates the data from the two data frames together, and we have joined them into one table that contains both the data about the mice as well as the expression of their genes.</span>
<span id="cb68-320"><a href="#cb68-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-321"><a href="#cb68-321" aria-hidden="true" tabindex="-1"></a>With the <span class="in">`left_join()`</span>, if there is a <span class="in">`id_num`</span> value that exists in the right data frame but not found in the left, then that data will not be present in the joined table.</span>
<span id="cb68-322"><a href="#cb68-322" aria-hidden="true" tabindex="-1"></a>For any <span class="in">`id_num`</span> that appears only in the left but not the right data frame, the values in the newly created columns will be <span class="in">`NA`</span>.</span>
<span id="cb68-323"><a href="#cb68-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-324"><a href="#cb68-324" aria-hidden="true" tabindex="-1"></a>::: callout-caution</span>
<span id="cb68-325"><a href="#cb68-325" aria-hidden="true" tabindex="-1"></a>If the values from the left data frame matches to multiple rows of the column in the right data frame, the <span class="in">`left_join()`</span> will duplicate the data from the left data frame for each match to the right.</span>
<span id="cb68-326"><a href="#cb68-326" aria-hidden="true" tabindex="-1"></a>This can cause issues with downstream summarisation if not carefully considered.</span>
<span id="cb68-327"><a href="#cb68-327" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb68-328"><a href="#cb68-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-329"><a href="#cb68-329" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Joining with mismatched column names</span></span>
<span id="cb68-330"><a href="#cb68-330" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-331"><a href="#cb68-331" aria-hidden="true" tabindex="-1"></a>Often the column containing the matching information has different names in different data frames.</span>
<span id="cb68-332"><a href="#cb68-332" aria-hidden="true" tabindex="-1"></a>For example what might be called "id_num" in one data frame could also be called "mouse_id" in another data frame.</span>
<span id="cb68-333"><a href="#cb68-333" aria-hidden="true" tabindex="-1"></a>In those cases the <span class="in">`by`</span> argument of <span class="in">`left_join()`</span> can be formatted to let the function know which column in the left data frame matches to which column on the right.</span>
<span id="cb68-334"><a href="#cb68-334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-335"><a href="#cb68-335" aria-hidden="true" tabindex="-1"></a>We will demonstrate this by renaming the <span class="in">`id_num`</span> column in the <span class="in">`m_dose`</span> to <span class="in">`mouse_id`</span> and using it to perform the join instead.</span>
<span id="cb68-336"><a href="#cb68-336" aria-hidden="true" tabindex="-1"></a>We will use the <span class="in">`join_by(mouse_id == id_num)`</span> helper function for the <span class="in">`by`</span> argument to specify the different columns we wish to join by.</span>
<span id="cb68-337"><a href="#cb68-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-340"><a href="#cb68-340" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-341"><a href="#cb68-341" aria-hidden="true" tabindex="-1"></a><span class="co"># rename id_num to mouse_id</span></span>
<span id="cb68-342"><a href="#cb68-342" aria-hidden="true" tabindex="-1"></a>m_dose_data_new <span class="ot">&lt;-</span> m_dose <span class="sc">%&gt;%</span></span>
<span id="cb68-343"><a href="#cb68-343" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">mouse_id =</span> id_num)</span>
<span id="cb68-344"><a href="#cb68-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-345"><a href="#cb68-345" aria-hidden="true" tabindex="-1"></a>m_dose_data_new</span>
<span id="cb68-346"><a href="#cb68-346" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-347"><a href="#cb68-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-350"><a href="#cb68-350" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-351"><a href="#cb68-351" aria-hidden="true" tabindex="-1"></a><span class="co"># perform left join by matching mouse_id of the left data frame to id_num of the right data frame</span></span>
<span id="cb68-352"><a href="#cb68-352" aria-hidden="true" tabindex="-1"></a>m_joined_data <span class="ot">&lt;-</span> <span class="fu">left_join</span>(</span>
<span id="cb68-353"><a href="#cb68-353" aria-hidden="true" tabindex="-1"></a>  m_dose_data_new, m_exp_rep_summed_wide,</span>
<span id="cb68-354"><a href="#cb68-354" aria-hidden="true" tabindex="-1"></a>  <span class="at">by =</span> <span class="fu">join_by</span>(mouse_id <span class="sc">==</span> id_num)</span>
<span id="cb68-355"><a href="#cb68-355" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb68-356"><a href="#cb68-356" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>() <span class="co"># keep only rows without NA</span></span>
<span id="cb68-357"><a href="#cb68-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-358"><a href="#cb68-358" aria-hidden="true" tabindex="-1"></a>m_joined_data</span>
<span id="cb68-359"><a href="#cb68-359" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-360"><a href="#cb68-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-361"><a href="#cb68-361" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb68-362"><a href="#cb68-362" aria-hidden="true" tabindex="-1"></a>When joining data frames with mismatched column names, the column name from the left data frame argument will be retained for the joined result.</span>
<span id="cb68-363"><a href="#cb68-363" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb68-364"><a href="#cb68-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-365"><a href="#cb68-365" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb68-366"><a href="#cb68-366" aria-hidden="true" tabindex="-1"></a>An older syntax exists for the <span class="in">`by`</span> argument that would look like <span class="in">`by = c("mouse_id" = "id_num")`</span>.</span>
<span id="cb68-367"><a href="#cb68-367" aria-hidden="true" tabindex="-1"></a>This does the same thing as what is shown above but the documentation now recommends the new syntax which allows for flexible options like joining to the closest.</span>
<span id="cb68-368"><a href="#cb68-368" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb68-369"><a href="#cb68-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-370"><a href="#cb68-370" aria-hidden="true" tabindex="-1"></a><span class="fu">## Visualising &amp; testing expression between strains</span></span>
<span id="cb68-371"><a href="#cb68-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-372"><a href="#cb68-372" aria-hidden="true" tabindex="-1"></a>With the joined data we now have more information about the mice in addition to the expression data.</span>
<span id="cb68-373"><a href="#cb68-373" aria-hidden="true" tabindex="-1"></a>So we can check if there is any significant difference in gene expressions between mouse strains.</span>
<span id="cb68-374"><a href="#cb68-374" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-377"><a href="#cb68-377" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-378"><a href="#cb68-378" aria-hidden="true" tabindex="-1"></a>m_joined_data <span class="sc">%&gt;%</span> </span>
<span id="cb68-379"><a href="#cb68-379" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>( <span class="fu">aes</span>(<span class="at">x =</span> mouse_strain, <span class="at">y =</span> Th, <span class="at">fill =</span> mouse_strain)) <span class="sc">+</span></span>
<span id="cb68-380"><a href="#cb68-380" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>()</span>
<span id="cb68-381"><a href="#cb68-381" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-382"><a href="#cb68-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-383"><a href="#cb68-383" aria-hidden="true" tabindex="-1"></a>If we wanted to plot all 3 genes, one way is to make 3 different plots, but we can also use facets.</span>
<span id="cb68-384"><a href="#cb68-384" aria-hidden="true" tabindex="-1"></a>However to use facets we would need to have the faceting variable in a column, and currently our gene names are in the column names.</span>
<span id="cb68-385"><a href="#cb68-385" aria-hidden="true" tabindex="-1"></a>So we must use <span class="in">`pivot_longer()`</span>.</span>
<span id="cb68-386"><a href="#cb68-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-389"><a href="#cb68-389" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-390"><a href="#cb68-390" aria-hidden="true" tabindex="-1"></a>m_joined_data_long <span class="ot">&lt;-</span> m_joined_data <span class="sc">%&gt;%</span></span>
<span id="cb68-391"><a href="#cb68-391" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(Th, Prlh, Hprt1), <span class="at">names_to =</span> <span class="st">"gene"</span>, <span class="at">values_to =</span> <span class="st">"expression"</span>)</span>
<span id="cb68-392"><a href="#cb68-392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-393"><a href="#cb68-393" aria-hidden="true" tabindex="-1"></a>m_joined_data_long</span>
<span id="cb68-394"><a href="#cb68-394" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-395"><a href="#cb68-395" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-396"><a href="#cb68-396" aria-hidden="true" tabindex="-1"></a>Now that the gene variable is in a column, we can use this information to create our faceted plot.</span>
<span id="cb68-397"><a href="#cb68-397" aria-hidden="true" tabindex="-1"></a>We use <span class="in">`facet_wrap()`</span> with the <span class="in">`scales = "free_y"`</span> argument so each gene can have its own y-axis.</span>
<span id="cb68-398"><a href="#cb68-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-401"><a href="#cb68-401" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-402"><a href="#cb68-402" aria-hidden="true" tabindex="-1"></a>m_joined_data_long <span class="sc">%&gt;%</span> </span>
<span id="cb68-403"><a href="#cb68-403" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>( <span class="fu">aes</span>(<span class="at">x =</span> mouse_strain, <span class="at">y =</span> expression, <span class="at">fill =</span> mouse_strain)) <span class="sc">+</span></span>
<span id="cb68-404"><a href="#cb68-404" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>() <span class="sc">+</span></span>
<span id="cb68-405"><a href="#cb68-405" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>gene, <span class="at">scales =</span> <span class="st">"free_y"</span>)</span>
<span id="cb68-406"><a href="#cb68-406" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-407"><a href="#cb68-407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-408"><a href="#cb68-408" aria-hidden="true" tabindex="-1"></a>There doesn't appear to be clear differences in the expression of these three genes between the haplotypes.</span>
<span id="cb68-409"><a href="#cb68-409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-410"><a href="#cb68-410" aria-hidden="true" tabindex="-1"></a><span class="fu">### Simple linear model (discrete x)</span></span>
<span id="cb68-411"><a href="#cb68-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-412"><a href="#cb68-412" aria-hidden="true" tabindex="-1"></a>To fit a line through the gene expression values of each mouse genotype, and test for significance of the x\~y relationship, we can use a handy function <span class="co">[</span><span class="ot">`lm_test()`</span><span class="co">](https://github.com/bansell/tidyrstats?tab=readme-ov-file#lm_test)</span> from the <span class="in">`tidyrstats`</span> library.</span>
<span id="cb68-413"><a href="#cb68-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-416"><a href="#cb68-416" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-417"><a href="#cb68-417" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyrstats)</span>
<span id="cb68-418"><a href="#cb68-418" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-419"><a href="#cb68-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-420"><a href="#cb68-420" aria-hidden="true" tabindex="-1"></a>This function takes a long-format data frame as input, and the formula for the linear model we want to fit.</span>
<span id="cb68-421"><a href="#cb68-421" aria-hidden="true" tabindex="-1"></a>(We will read more about linear models in later chapters).</span>
<span id="cb68-422"><a href="#cb68-422" aria-hidden="true" tabindex="-1"></a>For this test, for each gene, we want to know whether the response (gene expression ; aka 'y'), changes significantly with the predictor (mouse strain ; aka 'x').</span>
<span id="cb68-423"><a href="#cb68-423" aria-hidden="true" tabindex="-1"></a>The model formula will be <span class="in">`y ~ x1`</span>, that is <span class="in">`expression ~ mouse_strain`</span>.</span>
<span id="cb68-424"><a href="#cb68-424" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-427"><a href="#cb68-427" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-428"><a href="#cb68-428" aria-hidden="true" tabindex="-1"></a>m_joined_data_long <span class="sc">%&gt;%</span> </span>
<span id="cb68-429"><a href="#cb68-429" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(gene) <span class="sc">%&gt;%</span> </span>
<span id="cb68-430"><a href="#cb68-430" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lm_test</span>(expression <span class="sc">~</span> mouse_strain) <span class="sc">%&gt;%</span> </span>
<span id="cb68-431"><a href="#cb68-431" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(term<span class="sc">!=</span><span class="st">'intercept'</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb68-432"><a href="#cb68-432" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(gene,term)</span>
<span id="cb68-433"><a href="#cb68-433" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb68-434"><a href="#cb68-434" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-435"><a href="#cb68-435" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-436"><a href="#cb68-436" aria-hidden="true" tabindex="-1"></a>Our intuition from the violin plots is correct.</span>
<span id="cb68-437"><a href="#cb68-437" aria-hidden="true" tabindex="-1"></a>None of the genes show significant differences in expression between the Black6 vs the reference (BALBC), nor the CD-1 vs the reference.</span>
<span id="cb68-438"><a href="#cb68-438" aria-hidden="true" tabindex="-1"></a>Note that all of the P values are <span class="sc">\&gt;</span> 0.05, meaning we dont have enough evidence to reject the null hypothesis.</span>
<span id="cb68-439"><a href="#cb68-439" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-440"><a href="#cb68-440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-441"><a href="#cb68-441" aria-hidden="true" tabindex="-1"></a><span class="fu">## Visualising &amp; testing relationships between dosage and genes</span></span>
<span id="cb68-442"><a href="#cb68-442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-443"><a href="#cb68-443" aria-hidden="true" tabindex="-1"></a>With the joined data frame we can also visualise the relationship between the drug dosage and the genes, separated by mouse strain.</span>
<span id="cb68-444"><a href="#cb68-444" aria-hidden="true" tabindex="-1"></a>We do this by using our <span class="in">`m_joined_data_long`</span> to plot with <span class="in">`facet_grid()`</span> specifying the layout of our grid.</span>
<span id="cb68-445"><a href="#cb68-445" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-448"><a href="#cb68-448" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-449"><a href="#cb68-449" aria-hidden="true" tabindex="-1"></a>m_joined_data_long <span class="sc">%&gt;%</span> </span>
<span id="cb68-450"><a href="#cb68-450" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> drug_dose_g, <span class="at">y =</span> expression, <span class="at">color =</span> mouse_strain)) <span class="sc">+</span></span>
<span id="cb68-451"><a href="#cb68-451" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb68-452"><a href="#cb68-452" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>) <span class="sc">+</span></span>
<span id="cb68-453"><a href="#cb68-453" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(<span class="at">cols =</span> <span class="fu">vars</span>(mouse_strain), <span class="at">rows =</span> <span class="fu">vars</span>(gene), <span class="at">scales =</span> <span class="st">"free"</span>)</span>
<span id="cb68-454"><a href="#cb68-454" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-455"><a href="#cb68-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-456"><a href="#cb68-456" aria-hidden="true" tabindex="-1"></a>Although very minor, there appears to be some effect of dosage on gene expression that is strain specific.</span>
<span id="cb68-457"><a href="#cb68-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-458"><a href="#cb68-458" aria-hidden="true" tabindex="-1"></a><span class="fu">### Simple linear model (continuous x)</span></span>
<span id="cb68-459"><a href="#cb68-459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-460"><a href="#cb68-460" aria-hidden="true" tabindex="-1"></a>Again, we can test this using <span class="in">`lm_test()`</span>, where y is <span class="in">`expression`</span>, and x is now the continuous variable <span class="in">`drug_dose_g`</span>.</span>
<span id="cb68-461"><a href="#cb68-461" aria-hidden="true" tabindex="-1"></a>We are testing the association for each gene in each mouse strain, by grouping the input data before <span class="in">`lm_test()`</span>.</span>
<span id="cb68-462"><a href="#cb68-462" aria-hidden="true" tabindex="-1"></a>Note this is different from correcting for gene and mouse strain as covariates, because the model only tests one gene-strain combination at a time.</span>
<span id="cb68-463"><a href="#cb68-463" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-466"><a href="#cb68-466" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-467"><a href="#cb68-467" aria-hidden="true" tabindex="-1"></a>m_joined_data_long <span class="sc">%&gt;%</span> </span>
<span id="cb68-468"><a href="#cb68-468" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(mouse_strain, gene) <span class="sc">%&gt;%</span> </span>
<span id="cb68-469"><a href="#cb68-469" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lm_test</span>(expression <span class="sc">~</span> drug_dose_g) <span class="sc">%&gt;%</span> </span>
<span id="cb68-470"><a href="#cb68-470" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(term<span class="sc">!=</span><span class="st">'intercept'</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb68-471"><a href="#cb68-471" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(mouse_strain, gene, p.value)</span>
<span id="cb68-472"><a href="#cb68-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-473"><a href="#cb68-473" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-474"><a href="#cb68-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-475"><a href="#cb68-475" aria-hidden="true" tabindex="-1"></a>Here we can see, in fact there is a significant association between Mousezempic drug dose and gene expression: within the Black 6 strain, for the Hprt1 gene.</span>
<span id="cb68-476"><a href="#cb68-476" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-477"><a href="#cb68-477" aria-hidden="true" tabindex="-1"></a>::: {.callout-note title="Stating statistical findings"}</span>
<span id="cb68-478"><a href="#cb68-478" aria-hidden="true" tabindex="-1"></a>From our summary table above, can you finish this sentence which is the formal statement of results from a linear model?</span>
<span id="cb68-479"><a href="#cb68-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-480"><a href="#cb68-480" aria-hidden="true" tabindex="-1"></a>*"In the Black 6 mouse strain, for every 1 gram increase in drug dose, on average, Hprt1 gene expression ..."*</span>
<span id="cb68-481"><a href="#cb68-481" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb68-482"><a href="#cb68-482" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-483"><a href="#cb68-483" aria-hidden="true" tabindex="-1"></a>You can read and work through more <span class="in">`lm_test()`</span> examples in the <span class="co">[</span><span class="ot">Statistics in R Primer</span><span class="co">](stats_primer.qmd)</span>.</span>
<span id="cb68-484"><a href="#cb68-484" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-485"><a href="#cb68-485" aria-hidden="true" tabindex="-1"></a><span class="fu">## Saving multiple objects</span></span>
<span id="cb68-486"><a href="#cb68-486" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-487"><a href="#cb68-487" aria-hidden="true" tabindex="-1"></a>We can save our newly transformed and joined data frames for easy access using the <span class="in">`save()`</span> command, which takes any number of environment objects, and a new <span class="in">`Rdata`</span> format file name as arguments:</span>
<span id="cb68-488"><a href="#cb68-488" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-491"><a href="#cb68-491" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-492"><a href="#cb68-492" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(</span>
<span id="cb68-493"><a href="#cb68-493" aria-hidden="true" tabindex="-1"></a>  m_exp_rep_summed,</span>
<span id="cb68-494"><a href="#cb68-494" aria-hidden="true" tabindex="-1"></a>  m_exp_rep_summed_wide,</span>
<span id="cb68-495"><a href="#cb68-495" aria-hidden="true" tabindex="-1"></a>  m_joined_data_long,</span>
<span id="cb68-496"><a href="#cb68-496" aria-hidden="true" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">'data_processed/mousezempic_cleaned_joined.Rdata'</span>)</span>
<span id="cb68-497"><a href="#cb68-497" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-498"><a href="#cb68-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-499"><a href="#cb68-499" aria-hidden="true" tabindex="-1"></a><span class="fu">## Another case study - WHO tuberculosis data</span></span>
<span id="cb68-500"><a href="#cb68-500" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-501"><a href="#cb68-501" aria-hidden="true" tabindex="-1"></a>We can look at another dataset containing tuberculosis (TB) cases broken down by year, country, age, gender, and diagnosis method.</span>
<span id="cb68-502"><a href="#cb68-502" aria-hidden="true" tabindex="-1"></a>The data comes from the <span class="co">[</span><span class="ot">2014 World Health Organization Global Tuberculosis Report</span><span class="co">](https://www.who.int/teams/global-tuberculosis-programme/data)</span> and is typical of many real-life datasets in terms of structure.</span>
<span id="cb68-503"><a href="#cb68-503" aria-hidden="true" tabindex="-1"></a>There are redundant columns, odd variable codes and missing values.</span>
<span id="cb68-504"><a href="#cb68-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-505"><a href="#cb68-505" aria-hidden="true" tabindex="-1"></a>We want to tidy it up using the tools we have learned in order to try to extract some information from this data.</span>
<span id="cb68-506"><a href="#cb68-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-509"><a href="#cb68-509" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-510"><a href="#cb68-510" aria-hidden="true" tabindex="-1"></a>who</span>
<span id="cb68-511"><a href="#cb68-511" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-512"><a href="#cb68-512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-513"><a href="#cb68-513" aria-hidden="true" tabindex="-1"></a>The first thing we notice is that <span class="in">`country`</span>, <span class="in">`iso2`</span> and <span class="in">`iso3`</span> all seem to be different encodings for the country.</span>
<span id="cb68-514"><a href="#cb68-514" aria-hidden="true" tabindex="-1"></a>The <span class="in">`year`</span> column is fine as a variable but the rest of the columns seem appear to be coding values in the column names.</span>
<span id="cb68-515"><a href="#cb68-515" aria-hidden="true" tabindex="-1"></a>There are numeric values in the cells, and since we know this dataset is supposed to contain case numbers, and no column clearly denotes this, we suspect that the values in the columns are case numbers with the column names being descriptive of the cases.</span>
<span id="cb68-516"><a href="#cb68-516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-517"><a href="#cb68-517" aria-hidden="true" tabindex="-1"></a>Since we believe that many of these column names are in fact encoding data, we want to <span class="in">`pivot_longer()`</span> to put that information in a column for further processing.</span>
<span id="cb68-518"><a href="#cb68-518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-521"><a href="#cb68-521" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-522"><a href="#cb68-522" aria-hidden="true" tabindex="-1"></a><span class="co"># pivot longer with columns 5 to 60, drop NA values</span></span>
<span id="cb68-523"><a href="#cb68-523" aria-hidden="true" tabindex="-1"></a>who_long <span class="ot">&lt;-</span> who <span class="sc">%&gt;%</span></span>
<span id="cb68-524"><a href="#cb68-524" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(</span>
<span id="cb68-525"><a href="#cb68-525" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> <span class="dv">5</span><span class="sc">:</span><span class="dv">60</span>, <span class="co"># alternatively contains("_")</span></span>
<span id="cb68-526"><a href="#cb68-526" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"description"</span>,</span>
<span id="cb68-527"><a href="#cb68-527" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">"cases"</span>,</span>
<span id="cb68-528"><a href="#cb68-528" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_drop_na =</span> <span class="cn">TRUE</span></span>
<span id="cb68-529"><a href="#cb68-529" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb68-530"><a href="#cb68-530" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-531"><a href="#cb68-531" aria-hidden="true" tabindex="-1"></a>who_long</span>
<span id="cb68-532"><a href="#cb68-532" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-533"><a href="#cb68-533" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-534"><a href="#cb68-534" aria-hidden="true" tabindex="-1"></a>It doesn't look like the values in <span class="in">`description`</span> are unique, so we can try to count them to get a sense of what we're dealing with.</span>
<span id="cb68-535"><a href="#cb68-535" aria-hidden="true" tabindex="-1"></a>It will show us the unique values as well as the counts for each value.</span>
<span id="cb68-536"><a href="#cb68-536" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-539"><a href="#cb68-539" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-540"><a href="#cb68-540" aria-hidden="true" tabindex="-1"></a>who_long <span class="sc">%&gt;%</span></span>
<span id="cb68-541"><a href="#cb68-541" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(description)</span>
<span id="cb68-542"><a href="#cb68-542" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-543"><a href="#cb68-543" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-544"><a href="#cb68-544" aria-hidden="true" tabindex="-1"></a>Some patterns emerge, and it maybe be possible to guess at exactly what these values mean.</span>
<span id="cb68-545"><a href="#cb68-545" aria-hidden="true" tabindex="-1"></a>For example we see the presence of "m" and "f" in the encoding which probably encodes sex.</span>
<span id="cb68-546"><a href="#cb68-546" aria-hidden="true" tabindex="-1"></a>Luckily we have a data dictionary ready from the source that tells us what this means.</span>
<span id="cb68-547"><a href="#cb68-547" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-548"><a href="#cb68-548" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>The first part tells us if the data is new or old, in this case all data is new.</span>
<span id="cb68-549"><a href="#cb68-549" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The next part describe the type of TB:</span>
<span id="cb68-550"><a href="#cb68-550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-551"><a href="#cb68-551" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`rel`</span> for relapse</span>
<span id="cb68-552"><a href="#cb68-552" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`ep`</span> for extrapulmonary</span>
<span id="cb68-553"><a href="#cb68-553" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`sn`</span> for pulmonary TB that cannot be diagnosed with a pulmonary smear (smear negative)</span>
<span id="cb68-554"><a href="#cb68-554" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`sp`</span> for pulmonary TB that can be diagnosed with a smear (smear positive)</span>
<span id="cb68-555"><a href="#cb68-555" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-556"><a href="#cb68-556" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>The sex of the patient. Males and females denoted by <span class="in">`m`</span> and <span class="in">`f`</span> respectively.</span>
<span id="cb68-557"><a href="#cb68-557" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>The 7 age groups the data is divided into:</span>
<span id="cb68-558"><a href="#cb68-558" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-559"><a href="#cb68-559" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`014`</span>: 0 - 14 years old</span>
<span id="cb68-560"><a href="#cb68-560" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`1524`</span>: 15 - 24 years old</span>
<span id="cb68-561"><a href="#cb68-561" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`2534`</span>: 25 - 34 years old</span>
<span id="cb68-562"><a href="#cb68-562" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`3544`</span>: 35 - 44 years old</span>
<span id="cb68-563"><a href="#cb68-563" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`4554`</span>: 45 - 54 years old</span>
<span id="cb68-564"><a href="#cb68-564" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`5564`</span>: 55 - 64 years old</span>
<span id="cb68-565"><a href="#cb68-565" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`65`</span>: 65 or older</span>
<span id="cb68-566"><a href="#cb68-566" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-567"><a href="#cb68-567" aria-hidden="true" tabindex="-1"></a><span class="fu">### String manupulation</span></span>
<span id="cb68-568"><a href="#cb68-568" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-569"><a href="#cb68-569" aria-hidden="true" tabindex="-1"></a>Armed with that knowledge we can begin to tidy up our dataset!</span>
<span id="cb68-570"><a href="#cb68-570" aria-hidden="true" tabindex="-1"></a>The first thing we will do is tidy up the <span class="in">`description`</span> column, we will do this with some string manipulation functions from the tidyverse.</span>
<span id="cb68-571"><a href="#cb68-571" aria-hidden="true" tabindex="-1"></a>These functions help us manipulate strings programatically and are essential in dealing with real-world data.</span>
<span id="cb68-572"><a href="#cb68-572" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-573"><a href="#cb68-573" aria-hidden="true" tabindex="-1"></a>Some useful functions for string manipulation include:</span>
<span id="cb68-574"><a href="#cb68-574" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-575"><a href="#cb68-575" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`str_replace(string, pattern, replacement)`</span>: replaces the pattern in the string with the replacement.</span>
<span id="cb68-576"><a href="#cb68-576" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`str_extract(string, pattern)`</span>: extracts the part of the string matching the pattern.</span>
<span id="cb68-577"><a href="#cb68-577" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`str_remove(string, pattern)`</span>: removes the part of the string matching the pattern.</span>
<span id="cb68-578"><a href="#cb68-578" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-579"><a href="#cb68-579" aria-hidden="true" tabindex="-1"></a>The full list of functions can be found in the <span class="co">[</span><span class="ot">stringr package documentation</span><span class="co">](https://stringr.tidyverse.org/reference/index.html)</span> and patterns can be written in very powerful ways with <span class="co">[</span><span class="ot">regular expressions</span><span class="co">](https://r4ds.had.co.nz/strings.html#matching-patterns-with-regular-expressions)</span>.</span>
<span id="cb68-580"><a href="#cb68-580" aria-hidden="true" tabindex="-1"></a>For our purposes we will just use simply fixed patterns.</span>
<span id="cb68-581"><a href="#cb68-581" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-582"><a href="#cb68-582" aria-hidden="true" tabindex="-1"></a>We saw previously that we can use <span class="in">`separate()`</span> to break apart on column into multiple columns.</span>
<span id="cb68-583"><a href="#cb68-583" aria-hidden="true" tabindex="-1"></a>However, we see there is a problem in this dataset where the cases that are marked as relapse are encoded <span class="in">`newrel`</span> which doesn't leave us with a separator to use.</span>
<span id="cb68-584"><a href="#cb68-584" aria-hidden="true" tabindex="-1"></a>We can remedy this by using a combination of <span class="in">`mutate()`</span> and <span class="in">`str_replace()`</span> to replace all instances of <span class="in">`newrel`</span> with <span class="in">`new_rel`</span>.</span>
<span id="cb68-585"><a href="#cb68-585" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-588"><a href="#cb68-588" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-589"><a href="#cb68-589" aria-hidden="true" tabindex="-1"></a>who_long <span class="ot">&lt;-</span> who_long <span class="sc">%&gt;%</span></span>
<span id="cb68-590"><a href="#cb68-590" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">description =</span> <span class="fu">str_replace</span>(description, <span class="st">"newrel"</span>, <span class="st">"new_rel"</span>))</span>
<span id="cb68-591"><a href="#cb68-591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-592"><a href="#cb68-592" aria-hidden="true" tabindex="-1"></a>who_long</span>
<span id="cb68-593"><a href="#cb68-593" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-594"><a href="#cb68-594" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-595"><a href="#cb68-595" aria-hidden="true" tabindex="-1"></a>Since every single observation contains only new cases and no old cases, the <span class="in">`new_`</span> portion of the encoding is redundant.</span>
<span id="cb68-596"><a href="#cb68-596" aria-hidden="true" tabindex="-1"></a>So we can remove it using <span class="in">`str_remove()`</span>.</span>
<span id="cb68-597"><a href="#cb68-597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-600"><a href="#cb68-600" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-601"><a href="#cb68-601" aria-hidden="true" tabindex="-1"></a>who_long <span class="ot">&lt;-</span> who_long <span class="sc">%&gt;%</span></span>
<span id="cb68-602"><a href="#cb68-602" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">description =</span> <span class="fu">str_remove</span>(description, <span class="st">"new_"</span>))</span>
<span id="cb68-603"><a href="#cb68-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-604"><a href="#cb68-604" aria-hidden="true" tabindex="-1"></a>who_long</span>
<span id="cb68-605"><a href="#cb68-605" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-606"><a href="#cb68-606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-607"><a href="#cb68-607" aria-hidden="true" tabindex="-1"></a>Now one last problem remains, there is no separator between the sex and age group.</span>
<span id="cb68-608"><a href="#cb68-608" aria-hidden="true" tabindex="-1"></a>So we can once again use <span class="in">`str_replace()`</span> to edit in an underscore.</span>
<span id="cb68-609"><a href="#cb68-609" aria-hidden="true" tabindex="-1"></a>Note that we use <span class="in">`"_f"`</span> and <span class="in">`"_m"`</span> as our patterns as a precaution in case the letters naturally occur elsewhere in the encoding.</span>
<span id="cb68-610"><a href="#cb68-610" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-613"><a href="#cb68-613" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-614"><a href="#cb68-614" aria-hidden="true" tabindex="-1"></a>who_long <span class="ot">&lt;-</span> who_long <span class="sc">%&gt;%</span></span>
<span id="cb68-615"><a href="#cb68-615" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">description =</span> <span class="fu">str_replace</span>(description, <span class="st">"_f"</span>, <span class="st">"_f_"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb68-616"><a href="#cb68-616" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">description =</span> <span class="fu">str_replace</span>(description, <span class="st">"_m"</span>, <span class="st">"_m_"</span>))</span>
<span id="cb68-617"><a href="#cb68-617" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-618"><a href="#cb68-618" aria-hidden="true" tabindex="-1"></a>who_long</span>
<span id="cb68-619"><a href="#cb68-619" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-620"><a href="#cb68-620" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-621"><a href="#cb68-621" aria-hidden="true" tabindex="-1"></a>With that we are finally able to separate the data into individual columns with <span class="in">`separate()`</span>.</span>
<span id="cb68-622"><a href="#cb68-622" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-625"><a href="#cb68-625" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-626"><a href="#cb68-626" aria-hidden="true" tabindex="-1"></a>who_separated <span class="ot">&lt;-</span> who_long <span class="sc">%&gt;%</span></span>
<span id="cb68-627"><a href="#cb68-627" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(description, <span class="at">into =</span> <span class="fu">c</span>(<span class="st">"type"</span>, <span class="st">"sex"</span>, <span class="st">"age_group"</span>), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb68-628"><a href="#cb68-628" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-629"><a href="#cb68-629" aria-hidden="true" tabindex="-1"></a>who_separated</span>
<span id="cb68-630"><a href="#cb68-630" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-631"><a href="#cb68-631" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-632"><a href="#cb68-632" aria-hidden="true" tabindex="-1"></a>Because the columns <span class="in">`iso2`</span> and <span class="in">`iso3`</span> contain redundant information about the country, we can remove them, but also retain them for reference as we may want to know what the short forms of the countries are when relating it other tables.</span>
<span id="cb68-633"><a href="#cb68-633" aria-hidden="true" tabindex="-1"></a>To do so we can <span class="in">`select()`</span> the columns containing this information, then use <span class="in">`distinct()`</span> to keep only unique rows.</span>
<span id="cb68-634"><a href="#cb68-634" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-637"><a href="#cb68-637" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-638"><a href="#cb68-638" aria-hidden="true" tabindex="-1"></a>country_codes <span class="ot">&lt;-</span> who_separated <span class="sc">%&gt;%</span></span>
<span id="cb68-639"><a href="#cb68-639" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(country, iso2, iso3) <span class="sc">%&gt;%</span></span>
<span id="cb68-640"><a href="#cb68-640" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>()</span>
<span id="cb68-641"><a href="#cb68-641" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-642"><a href="#cb68-642" aria-hidden="true" tabindex="-1"></a>country_codes</span>
<span id="cb68-643"><a href="#cb68-643" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-644"><a href="#cb68-644" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-645"><a href="#cb68-645" aria-hidden="true" tabindex="-1"></a>With that we can safely remove the redundant information from our data and end up with a tidy data frame.</span>
<span id="cb68-646"><a href="#cb68-646" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-649"><a href="#cb68-649" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-650"><a href="#cb68-650" aria-hidden="true" tabindex="-1"></a>who_tidy <span class="ot">&lt;-</span> who_separated <span class="sc">%&gt;%</span></span>
<span id="cb68-651"><a href="#cb68-651" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>iso2, <span class="sc">-</span>iso3)</span>
<span id="cb68-652"><a href="#cb68-652" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-653"><a href="#cb68-653" aria-hidden="true" tabindex="-1"></a>who_tidy</span>
<span id="cb68-654"><a href="#cb68-654" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-655"><a href="#cb68-655" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-656"><a href="#cb68-656" aria-hidden="true" tabindex="-1"></a>From there we can easily visualise various aspects of the data for analysis.</span>
<span id="cb68-657"><a href="#cb68-657" aria-hidden="true" tabindex="-1"></a>For example we can see what country had the highest number of total cases in the year 2010.</span>
<span id="cb68-658"><a href="#cb68-658" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-661"><a href="#cb68-661" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-662"><a href="#cb68-662" aria-hidden="true" tabindex="-1"></a>who_tidy <span class="sc">%&gt;%</span></span>
<span id="cb68-663"><a href="#cb68-663" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(year <span class="sc">==</span> <span class="dv">2010</span>) <span class="sc">%&gt;%</span></span>
<span id="cb68-664"><a href="#cb68-664" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">total_cases =</span> <span class="fu">sum</span>(cases), <span class="at">.by =</span> country) <span class="sc">%&gt;%</span></span>
<span id="cb68-665"><a href="#cb68-665" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(total_cases))</span>
<span id="cb68-666"><a href="#cb68-666" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-667"><a href="#cb68-667" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-668"><a href="#cb68-668" aria-hidden="true" tabindex="-1"></a>We can extract only the data from the top 10 countries based in 2010 cases and plot how their total case numbers between the years 2000 and 2010.</span>
<span id="cb68-669"><a href="#cb68-669" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-672"><a href="#cb68-672" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-673"><a href="#cb68-673" aria-hidden="true" tabindex="-1"></a>top_10_countries <span class="ot">&lt;-</span> who_tidy <span class="sc">%&gt;%</span></span>
<span id="cb68-674"><a href="#cb68-674" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(year <span class="sc">==</span> <span class="dv">2010</span>) <span class="sc">%&gt;%</span></span>
<span id="cb68-675"><a href="#cb68-675" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">total_cases =</span> <span class="fu">sum</span>(cases), <span class="at">.by =</span> country) <span class="sc">%&gt;%</span></span>
<span id="cb68-676"><a href="#cb68-676" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(total_cases)) <span class="sc">%&gt;%</span></span>
<span id="cb68-677"><a href="#cb68-677" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb68-678"><a href="#cb68-678" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-679"><a href="#cb68-679" aria-hidden="true" tabindex="-1"></a>top_10_countries</span>
<span id="cb68-680"><a href="#cb68-680" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-681"><a href="#cb68-681" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-684"><a href="#cb68-684" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb68-685"><a href="#cb68-685" aria-hidden="true" tabindex="-1"></a>who_tidy <span class="sc">%&gt;%</span></span>
<span id="cb68-686"><a href="#cb68-686" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(</span>
<span id="cb68-687"><a href="#cb68-687" aria-hidden="true" tabindex="-1"></a>    country <span class="sc">%in%</span> top_10_countries<span class="sc">$</span>country,</span>
<span id="cb68-688"><a href="#cb68-688" aria-hidden="true" tabindex="-1"></a>    year <span class="sc">&gt;=</span> <span class="dv">2000</span>,</span>
<span id="cb68-689"><a href="#cb68-689" aria-hidden="true" tabindex="-1"></a>    year <span class="sc">&lt;=</span> <span class="dv">2010</span></span>
<span id="cb68-690"><a href="#cb68-690" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb68-691"><a href="#cb68-691" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(</span>
<span id="cb68-692"><a href="#cb68-692" aria-hidden="true" tabindex="-1"></a>    <span class="at">total_cases =</span> <span class="fu">sum</span>(cases), <span class="at">.by =</span> <span class="fu">c</span>(country, year)</span>
<span id="cb68-693"><a href="#cb68-693" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb68-694"><a href="#cb68-694" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> year, <span class="at">y =</span> total_cases, <span class="at">col =</span> country)) <span class="sc">+</span></span>
<span id="cb68-695"><a href="#cb68-695" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb68-696"><a href="#cb68-696" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>()</span>
<span id="cb68-697"><a href="#cb68-697" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb68-698"><a href="#cb68-698" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-699"><a href="#cb68-699" aria-hidden="true" tabindex="-1"></a><span class="fu">## Summary</span></span>
<span id="cb68-700"><a href="#cb68-700" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-701"><a href="#cb68-701" aria-hidden="true" tabindex="-1"></a>Today we learned about the concept of "tidy data".</span>
<span id="cb68-702"><a href="#cb68-702" aria-hidden="true" tabindex="-1"></a>A particular form of tabular data that is particularly suitable for the tidyverse.</span>
<span id="cb68-703"><a href="#cb68-703" aria-hidden="true" tabindex="-1"></a>It contains:</span>
<span id="cb68-704"><a href="#cb68-704" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-705"><a href="#cb68-705" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>one variable per column</span>
<span id="cb68-706"><a href="#cb68-706" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>one observation per row</span>
<span id="cb68-707"><a href="#cb68-707" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>one value per cell</span>
<span id="cb68-708"><a href="#cb68-708" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-709"><a href="#cb68-709" aria-hidden="true" tabindex="-1"></a>We learned to reshape data frames using various functions powerful functions.</span>
<span id="cb68-710"><a href="#cb68-710" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-711"><a href="#cb68-711" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`pivot_wider()`</span> reduces the number of rows and move information into new columns.</span>
<span id="cb68-712"><a href="#cb68-712" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`pivot_longer()`</span> reduces the number of columns and increase the number of rows. Moving the column names into a new column.</span>
<span id="cb68-713"><a href="#cb68-713" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`separate()`</span> separates multiple values encoded in a single column into individual columns based on a separator character.</span>
<span id="cb68-714"><a href="#cb68-714" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-715"><a href="#cb68-715" aria-hidden="true" tabindex="-1"></a>We learned to combine data frames using the <span class="in">`left_join()`</span> function.</span>
<span id="cb68-716"><a href="#cb68-716" aria-hidden="true" tabindex="-1"></a>This allows us to merge data frames together that have common columns, matching up the rows according to the shared column.</span>
<span id="cb68-717"><a href="#cb68-717" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-718"><a href="#cb68-718" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`left_join(left, right, by = shared_column)`</span> takes two data frames and merges them based on the <span class="in">`shared_column`</span> column. The information from the right data frame is added to the left if there is a match within the shared column.</span>
<span id="cb68-719"><a href="#cb68-719" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`left_join(left, right, by = join_by(left_col_name == right_col_name))`</span> can be used to join using columns from two data frames that encode the same values but have different column names</span>
<span id="cb68-720"><a href="#cb68-720" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-721"><a href="#cb68-721" aria-hidden="true" tabindex="-1"></a>We also learned some useful string manipulation functions that help us clean up data columns</span>
<span id="cb68-722"><a href="#cb68-722" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-723"><a href="#cb68-723" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`str_replace()`</span> lets us replace patterns in a string</span>
<span id="cb68-724"><a href="#cb68-724" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span><span class="in">`str_remove()`</span> lets us remove patterns in a string</span>
<span id="cb68-725"><a href="#cb68-725" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-726"><a href="#cb68-726" aria-hidden="true" tabindex="-1"></a>Together these make up a very versatile toolkit for transforming messy data you find in the wild into a tidy, more useful format to summarise information or produce plots with.</span>
<span id="cb68-727"><a href="#cb68-727" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-728"><a href="#cb68-728" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- :::::::::::::::::::: {.callout-important title="Practice exercises"} --&gt;</span></span>
<span id="cb68-729"><a href="#cb68-729" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-730"><a href="#cb68-730" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- Try these practice questions to test your understanding --&gt;</span></span>
<span id="cb68-731"><a href="#cb68-731" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-732"><a href="#cb68-732" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- :::::::: question --&gt;</span></span>
<span id="cb68-733"><a href="#cb68-733" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-734"><a href="#cb68-734" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 1\. What does the following code do? --&gt;</span></span>
<span id="cb68-735"><a href="#cb68-735" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-736"><a href="#cb68-736" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ```{r} --&gt;</span></span>
<span id="cb68-737"><a href="#cb68-737" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-738"><a href="#cb68-738" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- #| eval: false --&gt;</span></span>
<span id="cb68-739"><a href="#cb68-739" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-740"><a href="#cb68-740" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- m_dose %&gt;% --&gt;</span></span>
<span id="cb68-741"><a href="#cb68-741" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-742"><a href="#cb68-742" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--   summarise( --&gt;</span></span>
<span id="cb68-743"><a href="#cb68-743" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-744"><a href="#cb68-744" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--     med_tail = median(tail_length_mm, na.rm = TRUE), --&gt;</span></span>
<span id="cb68-745"><a href="#cb68-745" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-746"><a href="#cb68-746" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--     .by = c(mouse_strain, sex)) %&gt;% --&gt;</span></span>
<span id="cb68-747"><a href="#cb68-747" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-748"><a href="#cb68-748" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--   pivot_wider(names_from = sex, values_from = med_tail) --&gt;</span></span>
<span id="cb68-749"><a href="#cb68-749" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-750"><a href="#cb68-750" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ``` --&gt;</span></span>
<span id="cb68-751"><a href="#cb68-751" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-752"><a href="#cb68-752" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::::::: choices --&gt;</span></span>
<span id="cb68-753"><a href="#cb68-753" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-754"><a href="#cb68-754" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: choice --&gt;</span></span>
<span id="cb68-755"><a href="#cb68-755" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-756"><a href="#cb68-756" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- Pivots data into a wide format where there is a column for each sex. --&gt;</span></span>
<span id="cb68-757"><a href="#cb68-757" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-758"><a href="#cb68-758" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: --&gt;</span></span>
<span id="cb68-759"><a href="#cb68-759" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-760"><a href="#cb68-760" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: {.choice .correct-choice} --&gt;</span></span>
<span id="cb68-761"><a href="#cb68-761" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-762"><a href="#cb68-762" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- Calculates the median tail length for each unique combination of `mouse_strain` and `sex` in the `m_dose` data frame, then pivots into a wide format where there is a column for each sex. --&gt;</span></span>
<span id="cb68-763"><a href="#cb68-763" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-764"><a href="#cb68-764" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: --&gt;</span></span>
<span id="cb68-765"><a href="#cb68-765" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-766"><a href="#cb68-766" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: choice --&gt;</span></span>
<span id="cb68-767"><a href="#cb68-767" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-768"><a href="#cb68-768" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- Calculates the median tail length for each unique combination of `mouse_strain` and `sex` in the `m_dose` data frame, then pivots into a wide format where there is a column for each mouse strain. --&gt;</span></span>
<span id="cb68-769"><a href="#cb68-769" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-770"><a href="#cb68-770" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: --&gt;</span></span>
<span id="cb68-771"><a href="#cb68-771" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-772"><a href="#cb68-772" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: choice --&gt;</span></span>
<span id="cb68-773"><a href="#cb68-773" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-774"><a href="#cb68-774" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- It just gives an error --&gt;</span></span>
<span id="cb68-775"><a href="#cb68-775" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-776"><a href="#cb68-776" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: --&gt;</span></span>
<span id="cb68-777"><a href="#cb68-777" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-778"><a href="#cb68-778" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::::::: --&gt;</span></span>
<span id="cb68-779"><a href="#cb68-779" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-780"><a href="#cb68-780" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- :::::::: --&gt;</span></span>
<span id="cb68-781"><a href="#cb68-781" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-782"><a href="#cb68-782" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 2\. I have run the following code to create a new column in the `m_dose` data frame that gives the weight of the mice at the end of the experiment. --&gt;</span></span>
<span id="cb68-783"><a href="#cb68-783" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-784"><a href="#cb68-784" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ```{r} --&gt;</span></span>
<span id="cb68-785"><a href="#cb68-785" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-786"><a href="#cb68-786" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- m_dose %&gt;% --&gt;</span></span>
<span id="cb68-787"><a href="#cb68-787" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-788"><a href="#cb68-788" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--   # add a column for the weight at the end of the experiment --&gt;</span></span>
<span id="cb68-789"><a href="#cb68-789" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-790"><a href="#cb68-790" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--   mutate(final_weight_g = initial_weight_g - weight_lost_g) %&gt;% --&gt;</span></span>
<span id="cb68-791"><a href="#cb68-791" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-792"><a href="#cb68-792" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--   # select the relevant columns only --&gt;</span></span>
<span id="cb68-793"><a href="#cb68-793" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-794"><a href="#cb68-794" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--   select(id_num, initial_weight_g, final_weight_g) --&gt;</span></span>
<span id="cb68-795"><a href="#cb68-795" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-796"><a href="#cb68-796" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ``` --&gt;</span></span>
<span id="cb68-797"><a href="#cb68-797" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-798"><a href="#cb68-798" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- Which pivot function call would I use to take this data from a wide format (where there is a column for the final and initial weight) to a long format (where there is a row for each mouse and each weight measurement)? --&gt;</span></span>
<span id="cb68-799"><a href="#cb68-799" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-800"><a href="#cb68-800" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::::::: choices --&gt;</span></span>
<span id="cb68-801"><a href="#cb68-801" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-802"><a href="#cb68-802" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: choice --&gt;</span></span>
<span id="cb68-803"><a href="#cb68-803" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-804"><a href="#cb68-804" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- `pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "weight", values_to = "timepoint")` --&gt;</span></span>
<span id="cb68-805"><a href="#cb68-805" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-806"><a href="#cb68-806" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: --&gt;</span></span>
<span id="cb68-807"><a href="#cb68-807" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-808"><a href="#cb68-808" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: choice --&gt;</span></span>
<span id="cb68-809"><a href="#cb68-809" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-810"><a href="#cb68-810" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- `pivot_longer(cols = c(id_num, final_weight_g), names_to = "timepoint", values_to = "initial_weight_g")` --&gt;</span></span>
<span id="cb68-811"><a href="#cb68-811" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-812"><a href="#cb68-812" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: --&gt;</span></span>
<span id="cb68-813"><a href="#cb68-813" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-814"><a href="#cb68-814" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: choice --&gt;</span></span>
<span id="cb68-815"><a href="#cb68-815" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-816"><a href="#cb68-816" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- `pivot_wider(names_from = initial_weight_g, values_from = final_weight_g)` --&gt;</span></span>
<span id="cb68-817"><a href="#cb68-817" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-818"><a href="#cb68-818" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: --&gt;</span></span>
<span id="cb68-819"><a href="#cb68-819" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-820"><a href="#cb68-820" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: {.choice .correct-choice} --&gt;</span></span>
<span id="cb68-821"><a href="#cb68-821" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-822"><a href="#cb68-822" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- `pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")` --&gt;</span></span>
<span id="cb68-823"><a href="#cb68-823" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-824"><a href="#cb68-824" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: --&gt;</span></span>
<span id="cb68-825"><a href="#cb68-825" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-826"><a href="#cb68-826" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::::::: --&gt;</span></span>
<span id="cb68-827"><a href="#cb68-827" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-828"><a href="#cb68-828" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- :::::::: question --&gt;</span></span>
<span id="cb68-829"><a href="#cb68-829" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-830"><a href="#cb68-830" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 3\. Which of the following is NOT a valid way to join the `m_dose` data frame with the `m_exp` data frame based on the `id_num` column? --&gt;</span></span>
<span id="cb68-831"><a href="#cb68-831" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-832"><a href="#cb68-832" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::::::: choices --&gt;</span></span>
<span id="cb68-833"><a href="#cb68-833" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-834"><a href="#cb68-834" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: choice --&gt;</span></span>
<span id="cb68-835"><a href="#cb68-835" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-836"><a href="#cb68-836" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- `m_dose %&gt;% left_join(m_exp, by = "id_num")` --&gt;</span></span>
<span id="cb68-837"><a href="#cb68-837" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-838"><a href="#cb68-838" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: --&gt;</span></span>
<span id="cb68-839"><a href="#cb68-839" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-840"><a href="#cb68-840" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: choice --&gt;</span></span>
<span id="cb68-841"><a href="#cb68-841" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-842"><a href="#cb68-842" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- `left_join(m_dose, m_exp, by = "id_num")` --&gt;</span></span>
<span id="cb68-843"><a href="#cb68-843" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-844"><a href="#cb68-844" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: --&gt;</span></span>
<span id="cb68-845"><a href="#cb68-845" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-846"><a href="#cb68-846" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: {.choice .correct-choice} --&gt;</span></span>
<span id="cb68-847"><a href="#cb68-847" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-848"><a href="#cb68-848" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- `m_dose %&gt;% left_join(m_exp, .by = "id_num")` --&gt;</span></span>
<span id="cb68-849"><a href="#cb68-849" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-850"><a href="#cb68-850" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: --&gt;</span></span>
<span id="cb68-851"><a href="#cb68-851" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-852"><a href="#cb68-852" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: choice --&gt;</span></span>
<span id="cb68-853"><a href="#cb68-853" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-854"><a href="#cb68-854" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- `m_dose %&gt;% left_join(m_exp, by = ("id_num" = "id_num"))` --&gt;</span></span>
<span id="cb68-855"><a href="#cb68-855" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-856"><a href="#cb68-856" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::: --&gt;</span></span>
<span id="cb68-857"><a href="#cb68-857" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-858"><a href="#cb68-858" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- ::::::: --&gt;</span></span>
<span id="cb68-859"><a href="#cb68-859" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-860"><a href="#cb68-860" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- :::::::: --&gt;</span></span>
<span id="cb68-861"><a href="#cb68-861" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-862"><a href="#cb68-862" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- &lt;details&gt; --&gt;</span></span>
<span id="cb68-863"><a href="#cb68-863" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-864"><a href="#cb68-864" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- &lt;summary&gt;Solutions&lt;/summary&gt; --&gt;</span></span>
<span id="cb68-865"><a href="#cb68-865" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-866"><a href="#cb68-866" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 1\. The code first calculates the median tail length for each unique combination of `mouse_strain` and `sex` in the `m_dose` data frame, then pivots the data into a wide format where there is a column for each sex in the dataset (because of the argument `names_from = sex` ) --&gt;</span></span>
<span id="cb68-867"><a href="#cb68-867" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-868"><a href="#cb68-868" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 2\. The correct pivot function call to take the data from a wide format to a long format is `pivot_longer(cols = c(initial_weight_g, final_weight_g), names_to = "timepoint", values_to = "weight")`. This code tells R to pivot the `initial_weight_g` and `final_weight_g` columns into a long format, where there is a row for each mouse and each weight measurement. The `names_to` argument specifies to make a column called 'timepoint' that tells us whether the measurement is initial or final, and the `values_to` argument specifies the name of the new column that will contain these measurements. --&gt;</span></span>
<span id="cb68-869"><a href="#cb68-869" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-870"><a href="#cb68-870" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 3\. The line of code that is NOT a valid way to join the `m_dose` data frame with the `m_exp` data frame based on the `id_num` column is `m_dose %&gt;% left_join(m_exp, .by = "id_num")`. This line of code is incorrect because the `.by` argument is not used in the `left_join()` function (this can be confusing! it's `.by` when grouping by `by` when joining). The other options are valid ways to join the two data frames based on the `id_num` column: remember that we don't have to use pipes to join data frames, we can use the `left_join()` function directly, and we can use a named vector to specify the columns to join on (although here it's a bit redundant as the columns have the same name). --&gt;</span></span>
<span id="cb68-871"><a href="#cb68-871" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-872"><a href="#cb68-872" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- &lt;/details&gt; --&gt;</span></span>
<span id="cb68-873"><a href="#cb68-873" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-874"><a href="#cb68-874" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- :::::::::::::::::::: --&gt;</span></span>
<span id="cb68-875"><a href="#cb68-875" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-876"><a href="#cb68-876" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- summary - Reshaping data with `pivot_longer()` and `pivot_wider()` to change the structure of your data frame. --&gt;</span></span>
<span id="cb68-877"><a href="#cb68-877" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-878"><a href="#cb68-878" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- -   Combining data with `left_join()` to merge two data frames based on a common column. --&gt;</span></span>
<span id="cb68-879"><a href="#cb68-879" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-880"><a href="#cb68-880" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- questions etc --&gt;</span></span>
<span id="cb68-881"><a href="#cb68-881" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-882"><a href="#cb68-882" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 4.  Using the `m_dose` data frame, write R code to: --&gt;</span></span>
<span id="cb68-883"><a href="#cb68-883" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-884"><a href="#cb68-884" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--     a.  Make a data frame that shows the number of mice of each strain, in each replicate. --&gt;</span></span>
<span id="cb68-885"><a href="#cb68-885" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-886"><a href="#cb68-886" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--     b.  Pivot this data frame into a wide format to create a contingency table. --&gt;</span></span>
<span id="cb68-887"><a href="#cb68-887" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-888"><a href="#cb68-888" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--     c.  Pivot the wide data frame from (b) back into a long format. --&gt;</span></span>
<span id="cb68-889"><a href="#cb68-889" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-890"><a href="#cb68-890" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 5.  Let's say I have two data frames, `df1` and `df2`, that I want to join based a shared 'key' column, that is called 'key' in `df1` and 'item_key' in `df2`. Write R code to join these two data frames using the `left_join()` function. --&gt;</span></span>
<span id="cb68-891"><a href="#cb68-891" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-892"><a href="#cb68-892" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- solutions --&gt;</span></span>
<span id="cb68-893"><a href="#cb68-893" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-894"><a href="#cb68-894" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 4.  Here's how you could write R code to achieve the tasks: --&gt;</span></span>
<span id="cb68-895"><a href="#cb68-895" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-896"><a href="#cb68-896" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--     a.  `m_dose %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate))` --&gt;</span></span>
<span id="cb68-897"><a href="#cb68-897" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-898"><a href="#cb68-898" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--     b.  `m_dose %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate)) %&gt;% pivot_wider(names_from = replicate, values_from = n_mice)` --&gt;</span></span>
<span id="cb68-899"><a href="#cb68-899" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-900"><a href="#cb68-900" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!--     c.  `m_dose %&gt;% summarise(n_mice = n(), .by = c(mouse_strain, replicate)) %&gt;% pivot_wider(names_from = replicate, values_from = n_mice) %&gt;% pivot_longer(cols = starts_with("rep"), names_to = "replicate", values_to = "n_mice")` --&gt;</span></span>
<span id="cb68-901"><a href="#cb68-901" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-902"><a href="#cb68-902" aria-hidden="true" tabindex="-1"></a><span class="co">&lt;!-- 5.  To join the two data frames you could use `df1 %&gt;% left_join(df2, by = c("key" = "item_key"))` (with pipe) or `left_join(df1, df2, by = c("key" = "item_key"))` (without pipe). --&gt;</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




<script src="site_libs/quarto-contrib/alpine-3.12/alpine@3.12.min.js"></script>
</body></html>